<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>FeaturePlot — FeaturePlot • DESpace</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="FeaturePlot — FeaturePlot"><meta name="description" content="Plot spatial gene expression.
This function is a modified version of the FeaturePlot function from BayesSpace R package.
In comparison to the original BayesSpace function, this function allows plotting multiple genes simultaneously
and drawing an outline around a specified cluster."><meta property="og:description" content="Plot spatial gene expression.
This function is a modified version of the FeaturePlot function from BayesSpace R package.
In comparison to the original BayesSpace function, this function allows plotting multiple genes simultaneously
and drawing an outline around a specified cluster."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">DESpace</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.99.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/DSP.html">Differential Spatial Pattern between conditions</a></li>
    <li><a class="dropdown-item" href="../articles/SVG.html">A framework to discover Spatially Variable genes via spatial clusters</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/peicai/DESpace/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>FeaturePlot</h1>
      <small class="dont-index">Source: <a href="https://github.com/peicai/DESpace/blob/main/R/FeaturePlot.R" class="external-link"><code>R/FeaturePlot.R</code></a></small>
      <div class="d-none name"><code>FeaturePlot.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Plot spatial gene expression.
This function is a modified version of the <code>FeaturePlot</code> function from BayesSpace R package.
In comparison to the original BayesSpace function, this function allows plotting multiple genes simultaneously
and drawing an outline around a specified cluster.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">FeaturePlot</span><span class="op">(</span></span>
<span>  <span class="va">spe</span>,</span>
<span>  <span class="va">feature</span>,</span>
<span>  coordinates <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  concave_hull <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  sf_dim <span class="op">=</span> <span class="fl">200</span>,</span>
<span>  assay.type <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>  annotation_cluster <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  annotation_title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  platform <span class="op">=</span> <span class="st">"Visium"</span>,</span>
<span>  cluster_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cluster <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  legend_cluster <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  legend_exprs <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  diverging <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  low <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  high <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  mid <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  color <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  linewidth <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>  linecolor <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  label <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ncol <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  title <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  title_size <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  point_size <span class="op">=</span> <span class="fl">0.5</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-spe">spe<a class="anchor" aria-label="anchor" href="#arg-spe"></a></dt>
<dd><p>SpatialExperiment or SingleCellExperiment. If <code>feature</code> is specified and is a
string, it must exist as a row in the specified assay of <code>spe</code>.</p></dd>


<dt id="arg-feature">feature<a class="anchor" aria-label="anchor" href="#arg-feature"></a></dt>
<dd><p>Feature vector used to color each cell. May be the name of a
gene/row in an assay of <code>spe</code>, or a vector of continuous values.</p></dd>


<dt id="arg-coordinates">coordinates<a class="anchor" aria-label="anchor" href="#arg-coordinates"></a></dt>
<dd><p>Column names for the spatial coordinates of cells stored in <code>colData(spe)</code>.
If specified, these coordinates will be used. If not, the function defaults to using 'row' and 'col'
in <code>colData(spe)</code> if they exist. Otherwise, it will use <code>spatialCoords(spe)</code> if 'spe' is a
SpatialExperiment object and <code>spatialCoords(spe)</code> is not NULL.</p></dd>


<dt id="arg-concave-hull">concave_hull<a class="anchor" aria-label="anchor" href="#arg-concave-hull"></a></dt>
<dd><p>A logical value (TRUE or FALSE).
If TRUE, the function uses `ggforce::geom_mark_hull()` to outline cluster boundaries
(recommended for non-discontinuous clusters).
If FALSE, `sosta::reconstructShapeDensityImage()` is used for complex cluster shapes.
For Visium or ST platforms, `concave_hull` is automatically set to TRUE.</p></dd>


<dt id="arg-sf-dim">sf_dim<a class="anchor" aria-label="anchor" href="#arg-sf-dim"></a></dt>
<dd><p>A numeric value for the x-dimension of the reconstruction (default is 200).
A lower value speeds up computation but reduces accuracy.
Used only when `concave_hull` is FALSE.</p></dd>


<dt id="arg-assay-type">assay.type<a class="anchor" aria-label="anchor" href="#arg-assay-type"></a></dt>
<dd><p>String indicating which assay in <code>spe</code> the expression
vector should be taken from.</p></dd>


<dt id="arg-annotation-cluster">annotation_cluster<a class="anchor" aria-label="anchor" href="#arg-annotation-cluster"></a></dt>
<dd><p>A logical value (TRUE or FALSE).
If TRUE, annotated spatial clusters are plotted alongside expression plots.
If FALSE, clusters are not displayed.</p></dd>


<dt id="arg-annotation-title">annotation_title<a class="anchor" aria-label="anchor" href="#arg-annotation-title"></a></dt>
<dd><p>A character string for the title of the annotated spatial clusters.
Applied only when `annotation_cluster` is TRUE.</p></dd>


<dt id="arg-platform">platform<a class="anchor" aria-label="anchor" href="#arg-platform"></a></dt>
<dd><p>A character string specifying the spatial sequencing platform.
If "Visium" or "ST", a hexagonal spot layout will be used.
Otherwise, points will be plotted.</p></dd>


<dt id="arg-cluster-col">cluster_col<a class="anchor" aria-label="anchor" href="#arg-cluster-col"></a></dt>
<dd><p>Column name of spatial clusters in <code>colData(spe)</code>.</p></dd>


<dt id="arg-cluster">cluster<a class="anchor" aria-label="anchor" href="#arg-cluster"></a></dt>
<dd><p>Names of the spatial clusters used for drawing a boundary
around a group of points that belong to the specify cluster.
It can be NULL, "all"/"ALL", or a vector of cluster names.</p></dd>


<dt id="arg-legend-cluster">legend_cluster<a class="anchor" aria-label="anchor" href="#arg-legend-cluster"></a></dt>
<dd><p>A logical value. TRUE of FALSE,
indicating whether to plot the legend for the shaped clusters (TRUE), or not (FALSE).
Only used when 'cluster_col' and 'cluster' are specified.</p></dd>


<dt id="arg-legend-exprs">legend_exprs<a class="anchor" aria-label="anchor" href="#arg-legend-exprs"></a></dt>
<dd><p>A logical value.
TRUE of FALSE, indicating whether to plot the legend for
the expression level (TRUE), or not (FALSE).</p></dd>


<dt id="arg-diverging">diverging<a class="anchor" aria-label="anchor" href="#arg-diverging"></a></dt>
<dd><p>A logical value.
If TRUE, uses a diverging color gradient in <code>FeaturePlot</code> (e.g., for fold change).
If FALSE, uses a sequential gradient (e.g., for expression).</p></dd>


<dt id="arg-low-mid-high">low, mid, high<a class="anchor" aria-label="anchor" href="#arg-low-mid-high"></a></dt>
<dd><p>Optional hex codes for low, mid, and high values of the
color gradient used for continuous cell values.</p></dd>


<dt id="arg-color">color<a class="anchor" aria-label="anchor" href="#arg-color"></a></dt>
<dd><p>Optional hex code to set color of borders around cells. Set to
<code>NA</code> to remove borders.</p></dd>


<dt id="arg-linewidth">linewidth<a class="anchor" aria-label="anchor" href="#arg-linewidth"></a></dt>
<dd><p>The width of the boundary line around the cluster.
The default ('0.4') size of the boundary line is one.</p></dd>


<dt id="arg-linecolor">linecolor<a class="anchor" aria-label="anchor" href="#arg-linecolor"></a></dt>
<dd><p>The colors of the boundary lines around the cluster.
If unspecified, the default color scheme is used.</p></dd>


<dt id="arg-label">label<a class="anchor" aria-label="anchor" href="#arg-label"></a></dt>
<dd><p>A logical. TRUE of FALSE. Adding a label and an arrow pointing to a group.</p></dd>


<dt id="arg-ncol">ncol<a class="anchor" aria-label="anchor" href="#arg-ncol"></a></dt>
<dd><p>The dimensions of the grid to create.
By default, 1, if the length of feature equals to 1, and 3, otherwise.</p></dd>


<dt id="arg-title">title<a class="anchor" aria-label="anchor" href="#arg-title"></a></dt>
<dd><p>A logical. TRUE or FALSE. If true, the title name of each (subplot) is the gene name.</p></dd>


<dt id="arg-title-size">title_size<a class="anchor" aria-label="anchor" href="#arg-title-size"></a></dt>
<dd><p>Title font size.</p></dd>


<dt id="arg-point-size">point_size<a class="anchor" aria-label="anchor" href="#arg-point-size"></a></dt>
<dd><p>Point size.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Returns a ggplot object.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="svg_test.html">svg_test</a></code>, <code><a href="individual_svg.html">individual_svg</a></code>, <code><a href="top_results.html">top_results</a></code>, <code><a href="dsp_test.html">dsp_test</a></code>, <code><a href="individual_dsp.html">individual_dsp</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># load the input data:</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"LIBD_subset"</span>, package <span class="op">=</span> <span class="st">"DESpace"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># load pre-computed results (obtained via `svg_test`)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"results_svg_test"</span>, package <span class="op">=</span> <span class="st">"DESpace"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Visualize the gene expression of the top three genes</span></span></span>
<span class="r-in"><span><span class="va">feature</span> <span class="op">=</span> <span class="va">results_svg_test</span><span class="op">$</span><span class="va">gene_results</span><span class="op">$</span><span class="va">gene_id</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="fu">FeaturePlot</span><span class="op">(</span><span class="va">LIBD_subset</span>, <span class="va">feature</span>, coordinates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"array_row"</span>, <span class="st">"array_col"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>            ncol <span class="op">=</span> <span class="fl">3</span>, title <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="FeaturePlot-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Peiying Cai, Simone Tiberi.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

