<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Differential Spatial Pattern between conditions • DESpace</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Differential Spatial Pattern between conditions">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">DESpace</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/DSP.html">Differential Spatial Pattern between conditions</a></li>
    <li><a class="dropdown-item" href="../articles/SVG.html">A framework to discover Spatially Variable genes via spatial clusters</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/peicai/DESpace/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Differential Spatial Pattern between conditions</h1>
                        <h4 data-toc-skip class="author">Peiying
Cai</h4>
            <address class="author_afil">
      Institute for Molecular Life Sciences, University of Zurich,
SwitzerlandSIB Swiss Institute of Bioinformatics, University of Zurich,
Switzerland<br><a class="author_email" href="mailto:#"></a><a href="mailto:peiying.cai@uzh.ch" class="email">peiying.cai@uzh.ch</a>
      </address>
                              <h4 data-toc-skip class="author">Simone
Tiberi</h4>
            <address class="author_afil">
      Departiment of Statistical Sciences, University of Bologna,
Italy<br><a class="author_email" href="mailto:#"></a><a href="mailto:simone.tiberi@unibo.it" class="email">simone.tiberi@unibo.it</a>
      </address>
                  
            <h4 data-toc-skip class="date">10/31/2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/peicai/DESpace/blob/main/vignettes/DSP.Rmd" class="external-link"><code>vignettes/DSP.Rmd</code></a></small>
      <div class="d-none name"><code>DSP.Rmd</code></div>
    </div>

    
    
<hr>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><em>DESpace</em> is a framework for identifying spatially variable
genes (SVGs), a common task in spatial transcriptomics analyses, and
differential spatial variable pattern (DSP) genes, which identify
differences in spatial gene expression patterns across experimental
conditions.</p>
<p>By leveraging pre-annotated spatial clusters as summarized spatial
information, <em>DESpace</em> models gene expression with a negative
binomial (NB), via <em>edgeR</em> <span class="citation">(Robinson,
McCarthy, and Smyth 2010)</span>, with spatial clusters as covariates.
SV genes are then identified by testing the significance of spatial
clusters. For detailed guidance on detecting SVGs with <em>DESpace</em>,
refer to <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/DESpace/inst/doc/DESpace.html" class="external-link"><em>SVGs
vignettes</em></a>.</p>
<p>For multi-sample, multi-condition datasets, again we fit a NB model
via <em>edgeR</em> <span class="citation">(Robinson, McCarthy, and Smyth
2010)</span>, but this time we use spatial clusters, conditions and
their interactions as covariates. DSP genes are then identified by
testing the interaction between spatial clusters and conditions.
Notably, this framework can identify differences also between more than
2 groups. This vignette will demonstrate how to perform DSP
analyses.</p>
</div>
<div class="section level2">
<h2 id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/peicai/DESpace" class="external-link">DESpace</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/peicai/muSpaData" class="external-link">muSpaData</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hadley/reshape" class="external-link">reshape2</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">splines</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a>
</h2>
<p>As an example dataset, we consider a multi-sample, multi-group
spatially resolved transcriptomics data - the Stereo-seq dataset of
taxolotl telencephalon brain regeneration stages <span class="citation">(Wei et al. 2022)</span>. The dataset includes axolotl
brain tissues collected at various days post-injury (DPI): 2 (3
sections), 5 (3 sections), 10 (3 sections), 15 (4 sections), 20 (3
sections), 30 (1 section) and 60 (1 section), after the removal of a
reproducible portion of dorsal pallium in left telencephalic hemisphere
of axolotl. The original dataset is available for download via <a href="https://db.cngb.org/stomics/artista/download/" class="external-link"><em>STOmicsDB</em></a>,
and the processed dataset (including spatial clusters) can be accessed
via <a href="https://github.com/peicai/muSpaData" class="external-link"><em>muSpaData</em></a>
ExperimentHub package.</p>
<div class="section level3">
<h3 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h3>
<p>Here, we use a subset of the original data, consisting of three
distinct regeneration stages: 2, 10 and 20 DPI, with two sections for
each stage.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the small example data</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/muSpaData/man/Wei22_example.html" class="external-link">Wei22_example</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># The following columns from colData(spe) are specified:</span></span>
<span><span class="va">coordinates</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sdimx"</span>, <span class="st">"sdimy"</span><span class="op">)</span> <span class="co"># coordinates of cells</span></span>
<span><span class="va">spatial_cluster</span> <span class="op">&lt;-</span> <span class="st">'Banksy_smooth'</span> <span class="co"># Banksy spatial clusters</span></span>
<span><span class="va">condition_col</span> <span class="op">&lt;-</span> <span class="st">'condition'</span>       <span class="co"># regeneration time phases</span></span>
<span><span class="va">sample_col</span> <span class="op">&lt;-</span> <span class="st">'sample_id'</span>          <span class="co"># tissue section id</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 5 columns</span></span>
<span><span class="co">##                    sample_id condition Banksy_smooth     sdimx     sdimy</span></span>
<span><span class="co">##                     &lt;factor&gt;  &lt;factor&gt;      &lt;factor&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## CELL.17879.10DPI_1   10DPI_1     10DPI             4         1      2406</span></span>
<span><span class="co">## CELL.17922.10DPI_1   10DPI_1     10DPI             4        10      2372</span></span>
<span><span class="co">## CELL.17966.10DPI_1   10DPI_1     10DPI             4        29      3090</span></span>
<span><span class="co">## CELL.17976.10DPI_1   10DPI_1     10DPI             4        33      3139</span></span>
<span><span class="co">## CELL.17987.10DPI_1   10DPI_1     10DPI             4        33      2267</span></span>
<span><span class="co">## CELL.17988.10DPI_1   10DPI_1     10DPI             4        37      2791</span></span></code></pre>
<p>The spatial tissues of each sample were annotated via Banksy <span class="citation">(Singhal et al. 2024)</span>, classifying cells into
five clusters. These cluster annotations are stored in the
<code>Banksy_smooth</code> column of colData. Additionally, the columns
<code>sdimx</code> and <code>sdimy</code> contain the spatial
coordinates of the cells, while the <code>condition</code> column
specifies the group (i.e., stage) each cell belongs to.</p>
</div>
<div class="section level3">
<h3 id="quality-controlfiltering">Quality control/filtering<a class="anchor" aria-label="anchor" href="#quality-controlfiltering"></a>
</h3>
<p>Quality control (QC) procedures at the cell and gene level aim to
remove both low-quality cells, and lowly abundant genes. For QC, we
adhere to the instructions from “Orchestrating Spatially Resolved
Transcriptomics Analysis with Bioconductor” (<a href="https://lmweber.org/OSTA-book/quality-control.html" class="external-link"><em>OSTA</em></a>).
Library size and UMI counts are used to identify low-quality cells.
Then, we discard lowly abundant genes that are detected in fewer than 20
cells. R scripts for performing quality control on this example dataset
can be found in <a href="https://github.com/peicai/muSpaData/blob/main/inst/scripts/make-data.R#L94-L130" class="external-link"><em>muSpaData
R scripts</em></a>.</p>
</div>
<div class="section level3">
<h3 id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h3>
<p>This framework relies on spatial clusters being accessible and
successfully summarizing the primary spatial characteristics of the
data. In most datasets, these spatial features are either accessible or
can be generated with spatial clustering algorithms.</p>
<div class="section level4">
<h4 id="manual-annotation">Manual annotation<a class="anchor" aria-label="anchor" href="#manual-annotation"></a>
</h4>
<p>If the manual annotation (e.g., annotated by a pathologist) for each
sample is available, we can combine all samples and use manual
annotations directly. Note that cluster labels must be consistent across
samples (i.e., cluster 1 in sample 1 should represent the same tissue as
cluster 1 in sample 2).</p>
</div>
<div class="section level4">
<h4 id="spatially-resolved-multi-sample-clustering">Spatially resolved (multi-sample) clustering<a class="anchor" aria-label="anchor" href="#spatially-resolved-multi-sample-clustering"></a>
</h4>
<p>If manual annotations are not available, we can use spatially
resolved clustering tools. These methods, by jointly employing spatial
coordinates and gene expression data, enable obtaining spatial
clusters.</p>
<p>Among others, BayesSpace <span class="citation">(Zhao et al.
2021)</span> and Banksy <span class="citation">(Singhal et al.
2024)</span> allow jointly clustering multiple samples. In particular
each tool has a specific vignettes for multi-sample clustering: <a href="https://edward130603.github.io/BayesSpace/articles/joint_clustering.html" class="external-link"><em>BayesSpace
vignettes</em></a>, and <a href="https://prabhakarlab.github.io/Banksy/articles/multi-sample.html" class="external-link"><em>Banksy
vignettes</em></a>.</p>
<p>Details on applying Banksy joint clustering to this example dataset
can also be found in <a href="https://github.com/peicai/muSpaData/blob/main/inst/scripts/make-data.R#L131-L230" class="external-link"><em>muSpaData
R scripts</em></a>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View Banksy clusters </span></span>
<span><span class="co"># The spatial cluster assignments are available in the `colData(spe)`</span></span>
<span><span class="va">CD</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">CD</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">sdimx</span>, y <span class="op">=</span> <span class="va">sdimy</span>, color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">Banksy_smooth</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">sample_id</span>, scales <span class="op">=</span> <span class="st">'free'</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="cn">NULL</span>, title <span class="op">=</span> <span class="st">"Banksy Spatial Clusters"</span><span class="op">)</span></span></code></pre></div>
<p><img src="DSP_files/figure-html/view%20ARTISTA%20Banksy-1.png" width="700"></p>
<div class="section level5">
<h5 id="single-sample-clustering">Single sample clustering<a class="anchor" aria-label="anchor" href="#single-sample-clustering"></a>
</h5>
<p>In our benchmarks, we have noticed that, with both
<em>BayesSpace</em> and <em>Banksy</em>, joint spatial clustering of
multiple samples does not always yield more accurate results than
spatial clustering of individual samples. Therefore, if multi-sample
clustering fails, we suggest clustering individual samples (as described
in Section 3 <em>Individual sample</em> in the <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/DESpace/inst/doc/DESpace.html" class="external-link"><em>SVG
Vignette</em></a>) and manually matching cluster ids across samples, to
ensure that “cluster 1” always refers to the same spatial region in all
samples.</p>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="dsp-testing">DSP testing<a class="anchor" aria-label="anchor" href="#dsp-testing"></a>
</h2>
<p>Once we have spatial clusters, we can search for DSP between
conditions. Importantly, only clusters identified in all samples will be
analyzed.</p>
<div class="section level3">
<h3 id="gene-level-test">Gene-level test<a class="anchor" aria-label="anchor" href="#gene-level-test"></a>
</h3>
<p>Fit the model via <code>dsp_test</code> function. Parameter
<code>spe</code> specifies the input <code>SpatialExperiment</code> or
<code>SingleCellExperiment</code> object, while
<code>cluster_col</code>, <code>sample_col</code> and
<code>condition_col</code> define the column names in
<code>colData(spe)</code> for spatial clusters, sample ids, and
condition ids, respectively. Set <code>verbose</code> to
<code>TRUE</code> (default value) to view detailed statistics.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dsp_test.html">dsp_test</a></span><span class="op">(</span>spe <span class="op">=</span> <span class="va">spe</span>,</span>
<span>                    cluster_col <span class="op">=</span> <span class="va">spatial_cluster</span>,</span>
<span>                    sample_col <span class="op">=</span> <span class="va">sample_col</span>,</span>
<span>                    condition_col <span class="op">=</span> <span class="va">condition_col</span>,</span>
<span>                    verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Using 'dsp_test' for spatial variable pattern genes detection.</span></span></code></pre>
<pre><code><span><span class="co">## Filter low quality clusters:</span></span></code></pre>
<pre><code><span><span class="co">## Cluster levels to keep: 0, 1, 2, 3, 4</span></span></code></pre>
<pre><code><span><span class="co">## Design model: row names represent sample names, followed by underscores and cluster names.</span></span></code></pre>
<pre><code><span><span class="co">##          (Intercept) condition20DPI condition2DPI cluster_id1 cluster_id2</span></span>
<span><span class="co">## 2DPI_1_0           1              0             1           0           0</span></span>
<span><span class="co">## 2DPI_2_0           1              0             1           0           0</span></span>
<span><span class="co">##          cluster_id3 cluster_id4 condition20DPI:cluster_id1</span></span>
<span><span class="co">## 2DPI_1_0           0           0                          0</span></span>
<span><span class="co">## 2DPI_2_0           0           0                          0</span></span>
<span><span class="co">##          condition2DPI:cluster_id1 condition20DPI:cluster_id2</span></span>
<span><span class="co">## 2DPI_1_0                         0                          0</span></span>
<span><span class="co">## 2DPI_2_0                         0                          0</span></span>
<span><span class="co">##          condition2DPI:cluster_id2 condition20DPI:cluster_id3</span></span>
<span><span class="co">## 2DPI_1_0                         0                          0</span></span>
<span><span class="co">## 2DPI_2_0                         0                          0</span></span>
<span><span class="co">##          condition2DPI:cluster_id3 condition20DPI:cluster_id4</span></span>
<span><span class="co">## 2DPI_1_0                         0                          0</span></span>
<span><span class="co">## 2DPI_2_0                         0                          0</span></span>
<span><span class="co">##          condition2DPI:cluster_id4</span></span>
<span><span class="co">## 2DPI_1_0                         0</span></span>
<span><span class="co">## 2DPI_2_0                         0</span></span></code></pre>
<p>A list of results is returned, with the main results of interest
stored in the <code>gene_results</code> data frame. This frame contains
several columns, including gene names (<code>gene_id</code>), log2-fold
changes between groups (e.g,
<code>logFC.condition2DPI.cluster_id1</code>), average (across cells)
log-2 counts per million (<code>logCPM</code>), likelihood ratio test
statistics (<code>LR</code>), raw p-values (<code>PValue</code>) and
Benjamini-Hochberg adjusted p-values (<code>FDR</code>).</p>
<p>Specifically, the column <code>logFC.condition2DPI.cluster_id1</code>
represents the difference in the log2-fold change of gene expression
under <strong>2 DPI</strong> in <strong>cluster 1</strong> relative to
the baseline condition (<strong>10 DPI</strong>) and baseline cluster
(<strong>cluster 0</strong>).</p>
<p>In other words, we are testing whether the spatial structure of gene
expression (summarized by the clusters) differs between 2 and 10
DPI.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">gene_results</span>, <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                       gene_id logFC.condition20DPI.cluster_id1</span></span>
<span><span class="co">## AMEX60DD014721 AMEX60DD014721                      -0.07164635</span></span>
<span><span class="co">## AMEX60DD045083 AMEX60DD045083                       0.08745636</span></span>
<span><span class="co">##                logFC.condition2DPI.cluster_id1 logFC.condition20DPI.cluster_id2</span></span>
<span><span class="co">## AMEX60DD014721                      -0.3898630                       -0.8509553</span></span>
<span><span class="co">## AMEX60DD045083                       0.5417202                       -1.3379152</span></span>
<span><span class="co">##                logFC.condition2DPI.cluster_id2 logFC.condition20DPI.cluster_id3</span></span>
<span><span class="co">## AMEX60DD014721                       0.8526198                       -0.8327653</span></span>
<span><span class="co">## AMEX60DD045083                       1.1227184                        0.2479755</span></span>
<span><span class="co">##                logFC.condition2DPI.cluster_id3 logFC.condition20DPI.cluster_id4</span></span>
<span><span class="co">## AMEX60DD014721                      -0.9450822                        0.2229300</span></span>
<span><span class="co">## AMEX60DD045083                       1.2663910                       -0.9966468</span></span>
<span><span class="co">##                logFC.condition2DPI.cluster_id4   logCPM        LR       PValue</span></span>
<span><span class="co">## AMEX60DD014721                       0.2085998 9.344907 102.25832 1.473888e-18</span></span>
<span><span class="co">## AMEX60DD045083                      -0.7966179 7.505402  95.35246 3.791697e-17</span></span>
<span><span class="co">##                         FDR</span></span>
<span><span class="co">## AMEX60DD014721 7.369440e-15</span></span>
<span><span class="co">## AMEX60DD045083 9.479243e-14</span></span></code></pre>
<p>The second element of the results (a <code>DGEList</code> object
<code>estimated_y</code>) contains the estimated common dispersion.</p>
<p>The third and fourth element of the results (<code>DGEGLM</code> and
<code>DGELRT</code> objects) contain full statistics from
<code><a href="https://rdrr.io/pkg/edgeR/man/glmfit.html" class="external-link">edgeR::glmFit</a></code> and <code><a href="https://rdrr.io/pkg/edgeR/man/glmLRT.html" class="external-link">edgeR::glmLRT</a></code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">estimated_y</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">glmLrt</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">glmFit</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "DGEList"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "edgeR"</span></span></code></pre>
<pre><code><span><span class="co">## [1] "NULL"</span></span></code></pre>
<pre><code><span><span class="co">## [1] "DGEGLM"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "edgeR"</span></span></code></pre>
<p>Visualize the gene expression of the most significant genes with
<code><a href="../reference/FeaturePlot.html">FeaturePlot()</a></code>. Note that the gene names in vector
<code>feature</code>, should also appear in the count matrix’s row
names. Specifying the column names of spatial coordinates of spots is
only necessary when they are not named <code>row</code> and
<code>col</code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">CD</span><span class="op">$</span><span class="va">sample_id</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Identify the top DSP</span></span>
<span><span class="op">(</span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">results</span><span class="op">$</span><span class="va">gene_results</span><span class="op">$</span><span class="va">gene_id</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "AMEX60DD014721"</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract the gene_name by matching the gene_id</span></span>
<span><span class="op">(</span><span class="va">feature_name</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">gene_id</span><span class="op">[</span></span>
<span>  <span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">gene_name</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">feature</span></span>
<span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "ECM1"</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># generate a list of plots</span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">sample_ids</span>, <span class="kw">function</span><span class="op">(</span><span class="va">sample_id</span><span class="op">)</span> <span class="op">{</span></span>
<span>  </span>
<span>  <span class="co"># Subset spe for each sample</span></span>
<span>  <span class="va">spe_j</span> <span class="op">&lt;-</span> <span class="va">spe</span><span class="op">[</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">==</span> <span class="va">sample_id</span><span class="op">]</span></span>
<span>  </span>
<span>  <span class="co"># Create FeaturePlot for the sample</span></span>
<span>  <span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">spe_j</span>, <span class="va">feature</span>,</span>
<span>                      coordinates <span class="op">=</span> <span class="va">coordinates</span>,</span>
<span>                      platform <span class="op">=</span> <span class="st">"Stereo-seq"</span>, ncol <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                      diverging <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                      point_size <span class="op">=</span> <span class="fl">0.1</span>, legend_exprs <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"right"</span>,</span>
<span>          legend.key.size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="st">'cm'</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="va">sample_id</span><span class="op">)</span> </span>
<span>  </span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">plot</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>The spatial structure of gene expression changes across conditions,
transitioning from more localized patterns at earlier stages (2 and 10
DPI) to a broader distribution at a later stage (20 DPI).</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">combined_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plots</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="co"># common legend</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">'collect'</span><span class="op">)</span>  </span>
<span><span class="va">combined_plot</span></span></code></pre></div>
<p><img src="DSP_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="individual-cluster-test">Individual cluster test<a class="anchor" aria-label="anchor" href="#individual-cluster-test"></a>
</h3>
<p><em>DESpace</em> can also be used to reveal the specific areas of the
tissue affected by spatial variability; i.e., spatial clusters that are
particularly over/under abundant compared to the average across
conditions. Function <code><a href="../reference/individual_dsp.html">individual_dsp()</a></code> can be used to
identify DSP genes for each individual cluster. Parameters
<code>cluster_col</code>, <code>sample_col</code> and
<code>condition_col</code> indicate the column names in
<code>colData(spe)</code> for spatial clusters, sample ids, and
condition ids, respectively.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cluster_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/individual_dsp.html">individual_dsp</a></span><span class="op">(</span><span class="va">spe</span>,</span>
<span>                                  cluster_col <span class="op">=</span> <span class="va">spatial_cluster</span>,</span>
<span>                                  sample_col <span class="op">=</span> <span class="va">sample_col</span>,</span>
<span>                                  condition_col <span class="op">=</span> <span class="va">condition_col</span><span class="op">)</span></span></code></pre></div>
<p><code><a href="../reference/individual_dsp.html">individual_dsp()</a></code> returns a list containing the results
of the individual cluster tests. Similarly to above, the results for
each cluster are presented as a <code>data.fame</code>, where columns
contain gene names (<code>gene_id</code>), likelihood ratio test
statistics (<code>LR</code>), log2-fold changes (<code>logFC</code>),
raw p-values (<code>PValue</code>) and Benjamini-Hochberg adjusted
p-values (<code>FDR</code>).</p>
<p>Here, we present the top results for cluster 2.
<code>logFC.condition20DPI.cluster_id2</code> represents the interaction
between the 20 DPI condition and cluster 2. It compares the effect of 20
DPI in cluster 2 with its effect in all other clusters (i.e., all tissue
regions excluding cluster 2, which serves as the baseline). A positive
log-fold change value suggests that, the increase in gene expression in
<strong>cluster 2</strong> from 10 DPI (the baseline) to 20 DPI is
<strong>greater</strong> than the increase in gene expression in
<strong>all other clusters</strong> from 10 DPI to 20 DPI.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">cluster_results</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "list"</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cluster_results</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "0" "1" "2" "3" "4"</span></span></code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cluster_results</span><span class="op">$</span><span class="va">`2`</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                       gene_id logFC.condition20DPI.cluster_id2</span></span>
<span><span class="co">## AMEX60DD014721 AMEX60DD014721                       -0.5801491</span></span>
<span><span class="co">## AMEX60DD014991 AMEX60DD014991                        2.6942759</span></span>
<span><span class="co">## AMEX60DD055246 AMEX60DD055246                       -0.2661447</span></span>
<span><span class="co">## AMEX60DD045083 AMEX60DD045083                       -1.0631435</span></span>
<span><span class="co">##                logFC.condition2DPI.cluster_id2   logCPM       LR       PValue</span></span>
<span><span class="co">## AMEX60DD014721                       1.2700494 9.582333 76.16657 2.888278e-17</span></span>
<span><span class="co">## AMEX60DD014991                       3.0151238 7.330327 73.02536 1.389135e-16</span></span>
<span><span class="co">## AMEX60DD055246                      -2.2986026 5.661946 60.91208 5.930751e-14</span></span>
<span><span class="co">## AMEX60DD045083                       0.9065434 8.266790 59.19339 1.400618e-13</span></span>
<span><span class="co">##                         FDR</span></span>
<span><span class="co">## AMEX60DD014721 1.444139e-13</span></span>
<span><span class="co">## AMEX60DD014991 3.472837e-13</span></span>
<span><span class="co">## AMEX60DD055246 9.884585e-11</span></span>
<span><span class="co">## AMEX60DD045083 1.750772e-10</span></span></code></pre>
<p>Visualize the gene expression of the top gene for cluster 2.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># one of top DSPs for cluster 2</span></span>
<span><span class="op">(</span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cluster_results</span><span class="op">[[</span><span class="st">"2"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "AMEX60DD045083"</span></span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract the gene_name by matching the gene_id</span></span>
<span><span class="op">(</span><span class="va">feature_name</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">gene_id</span><span class="op">[</span></span>
<span>  <span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">gene_name</span> <span class="op">==</span> <span class="va">feature</span></span>
<span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "SFRP2"</span></span></code></pre>
<div class="section level4">
<h4 class="tabset" id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h4>
<div class="section level5">
<h5 id="abundance-trend">Abundance trend<a class="anchor" aria-label="anchor" href="#abundance-trend"></a>
</h5>
<p>One way is to plot the overall abundance of SFRP2 for each
cluster-sample combination. Under the null hypothesis, gene expression
changes across conditions are consistent across clusters.</p>
<p>The boxplots below show the average log-CPM for cluster 2 and for all
other clusters (excluding cluster 2) across different stages. In Cluster
2, the average abundance is highest at 2 DPI, then decreases at 10 DPI
and continues to drop at 20 DPI. In contrast, although there is a slight
decrease in abundance across other clusters, it remains relatively
constant overall.</p>
<details><summary>
Code
</summary><div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate log cpm</span></span>
<span><span class="va">cps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html" class="external-link">cpm</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">estimated_y</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">cps_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">cps</span><span class="op">)</span></span>
<span><span class="va">mdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    log_cpm <span class="op">=</span> <span class="va">cps</span><span class="op">[</span><span class="va">feature</span>, <span class="op">]</span> ,</span>
<span>    Banksy_smooth <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">".*_"</span>, <span class="st">""</span>, <span class="va">cps_name</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    day <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"([0-9]+)DPI.*"</span>, <span class="st">"\\1"</span>, <span class="va">cps_name</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    sample_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"(_[0-9]+)$"</span>, <span class="st">""</span>, <span class="va">cps_name</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">plt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">mdata</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">day</span><span class="op">)</span>, y <span class="op">=</span> <span class="va">log_cpm</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">Banksy_smooth</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">2</span>, width <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">Banksy_smooth</span> <span class="op">==</span> <span class="st">"2"</span>, </span>
<span>                                   <span class="st">"cluster 2"</span>, <span class="st">"non-cluster 2"</span><span class="op">)</span><span class="op">)</span>, </span>
<span>                 position <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_dodge.html" class="external-link">position_dodge</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html" class="external-link">scale_x_discrete</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">10</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#4DAF4A"</span>, <span class="st">"grey"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="va">feature_name</span>, x <span class="op">=</span> <span class="st">"Days post injury"</span>, </span>
<span>         y <span class="op">=</span> <span class="st">"log-2 counts per million (logCPM)"</span>, fill <span class="op">=</span> <span class="st">""</span>,</span>
<span>         color <span class="op">=</span> <span class="st">"Banksy cluster"</span><span class="op">)</span> <span class="op">+</span></span>
<span>     <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span></span></code></pre></div>
</details><div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># figure</span></span>
<span><span class="va">plt</span></span></code></pre></div>
<p><img src="DSP_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
<div class="section level5">
<h5 id="spatial-expression">Spatial expression<a class="anchor" aria-label="anchor" href="#spatial-expression"></a>
</h5>
<p>Alternatively, gene expression can be visualized in physical space
with <code><a href="../reference/FeaturePlot.html">FeaturePlot()</a></code>. A cluster outline drawn by specifying
the column names of clusters stored in <code>colData(spe)</code> and the
vector of cluster names via <code>cluster_col</code> and
<code>cluster</code>.</p>
<details><summary>
Code
</summary><div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># generate a list of FeaturePlots</span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">sample_ids</span>, <span class="kw">function</span><span class="op">(</span><span class="va">sample_id</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># Subset spe for each sample</span></span>
<span>    <span class="va">spe_j</span> <span class="op">&lt;-</span> <span class="va">spe</span><span class="op">[</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">==</span> <span class="va">sample_id</span><span class="op">]</span></span>
<span>    <span class="co"># Create FeaturePlot for the sample</span></span>
<span>    <span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">spe_j</span>, <span class="va">feature</span>, </span>
<span>                        cluster_col <span class="op">=</span> <span class="va">spatial_cluster</span>,</span>
<span>                        coordinates <span class="op">=</span> <span class="va">coordinates</span>, cluster <span class="op">=</span> <span class="st">'2'</span>,</span>
<span>                        platform <span class="op">=</span> <span class="st">"Stereo-seq"</span>,</span>
<span>                        diverging <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                        point_size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                        linewidth <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"right"</span>,</span>
<span>              legend.key.size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="st">'cm'</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="va">sample_id</span><span class="op">)</span> </span>
<span>  </span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">plot</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">combined_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plots</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="co"># common legend</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">'collect'</span><span class="op">)</span>  </span></code></pre></div>
</details><p>Again, the spatial structure of gene expression varies across groups;
in particular, at 2 and 10 DPI, abundance is higher in cluster 2
(outlined in the plot), compared to the rest of the tissue, while at 20
DPI abundance is more homogeneous.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># figure</span></span>
<span><span class="va">combined_plot</span> </span></code></pre></div>
<p><img src="DSP_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="smooth-splines-to-model-time">Smooth splines to model time<a class="anchor" aria-label="anchor" href="#smooth-splines-to-model-time"></a>
</h3>
<p><em>DESpace</em> offers a flexible framework that allows users to
create a custom design matrix. The default design matrix is
<code>model.matrix(~ condition * cluster)</code>. Below, we provide an
example of how to create a design matrix using piecewise-cubic splines
to account for the effect of time.</p>
<p>First, we create metadata associated with the samples and clusters.
For each cluster level, there are 3 time phases (i.e., <code>day</code>)
and 2 replicates (i.e., `rep``) for each time point.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># all combinations of sample and cluster</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/expand.grid.html" class="external-link">expand.grid</a></span><span class="op">(</span>sample_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">sample_id</span><span class="op">)</span>,</span>
<span>                        cluster <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">Banksy_smooth</span><span class="op">)</span></span>
<span>                        <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="co"># extract time point as 'day' from sample_id </span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>      day <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"DPI.*"</span>, <span class="st">""</span>, <span class="va">sample_id</span><span class="op">)</span><span class="op">)</span>,</span>
<span>      rep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">".*_"</span>, <span class="st">""</span>, <span class="va">sample_id</span><span class="op">)</span><span class="op">)</span> </span>
<span>      <span class="op">)</span></span>
<span><span class="va">metadata</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   sample_id cluster day rep</span></span>
<span><span class="co">## 1    2DPI_1       0   2   1</span></span>
<span><span class="co">## 2    2DPI_2       0   2   2</span></span>
<span><span class="co">## 3   10DPI_1       0  10   1</span></span></code></pre>
<p>Instead of treating time phases (e.g., 2 DPI, 10 DPI, 20 DPI) as a
categorical variable, we can model the time trend using a smooth spline
function. This can be achieved with the <code>ns(x, df)</code> function
from the <a href="https://stat.ethz.ch/R-manual/R-devel/library/splines/html/ns.html" class="external-link"><em>splines</em></a>
package. Here, <code>x</code> represents the predictor variable—time
phases (<code>day</code> in the <code>metadata</code>) in our case-and
<code>df</code> specifies the degrees of freedom, which determine the
total number of parameters in the <code><a href="https://rdrr.io/r/splines/ns.html" class="external-link">ns()</a></code> time model,
including the intercept.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">design_model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">cluster</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/splines/ns.html" class="external-link">ns</a></span><span class="op">(</span><span class="va">day</span>, df <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>, </span>
<span>                             data <span class="op">=</span> <span class="va">metadata</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">design_model</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">sample_id</span>, <span class="st">"_"</span>,</span>
<span>                                 <span class="va">metadata</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">design_model</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 30 15</span></span></code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">design_model</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           (Intercept) cluster1 cluster2 cluster3 cluster4 ns(day, df = 2)1</span></span>
<span><span class="co">## 2DPI_1_0            1        0        0        0        0        0.0000000</span></span>
<span><span class="co">## 2DPI_2_0            1        0        0        0        0        0.0000000</span></span>
<span><span class="co">## 10DPI_1_0           1        0        0        0        0        0.5513298</span></span>
<span><span class="co">##           ns(day, df = 2)2 cluster1:ns(day, df = 2)1 cluster2:ns(day, df = 2)1</span></span>
<span><span class="co">## 2DPI_1_0         0.0000000                         0                         0</span></span>
<span><span class="co">## 2DPI_2_0         0.0000000                         0                         0</span></span>
<span><span class="co">## 10DPI_1_0       -0.2274421                         0                         0</span></span>
<span><span class="co">##           cluster3:ns(day, df = 2)1 cluster4:ns(day, df = 2)1</span></span>
<span><span class="co">## 2DPI_1_0                          0                         0</span></span>
<span><span class="co">## 2DPI_2_0                          0                         0</span></span>
<span><span class="co">## 10DPI_1_0                         0                         0</span></span>
<span><span class="co">##           cluster1:ns(day, df = 2)2 cluster2:ns(day, df = 2)2</span></span>
<span><span class="co">## 2DPI_1_0                          0                         0</span></span>
<span><span class="co">## 2DPI_2_0                          0                         0</span></span>
<span><span class="co">## 10DPI_1_0                         0                         0</span></span>
<span><span class="co">##           cluster3:ns(day, df = 2)2 cluster4:ns(day, df = 2)2</span></span>
<span><span class="co">## 2DPI_1_0                          0                         0</span></span>
<span><span class="co">## 2DPI_2_0                          0                         0</span></span>
<span><span class="co">## 10DPI_1_0                         0                         0</span></span></code></pre>
<p>Fit the model via <code>dsp_test</code> function.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dsp_test.html">dsp_test</a></span><span class="op">(</span><span class="va">spe</span>,</span>
<span>                    design <span class="op">=</span> <span class="va">design_model</span>,</span>
<span>                    cluster_col <span class="op">=</span> <span class="va">spatial_cluster</span>,</span>
<span>                    sample_col <span class="op">=</span> <span class="va">sample_col</span>,</span>
<span>                    condition_col <span class="op">=</span> <span class="va">condition_col</span>,</span>
<span>                    verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># count significant DSP genes (at 5% FDR significance level)</span></span>
<span><span class="va">res_global</span> <span class="op">&lt;-</span> <span class="va">results</span><span class="op">$</span><span class="va">gene_results</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">res_global</span><span class="op">$</span><span class="va">FDR</span> <span class="op">&lt;=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## FALSE  TRUE </span></span>
<span><span class="co">##  4799   201</span></span></code></pre>
<p>To identify key spatial clusters where expression changes across
conditions, we apply the smooth spline with a single-cluster design.
Specifically, we convert the original Banksy clusters into two groups:
the target cluster and all other clusters. We then apply the same test
as in the global test above.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># example: testing for cluster 2</span></span>
<span><span class="co"># convert 5 Banksy clusters into 2 groups: cluster 2 vs. all other clusters</span></span>
<span><span class="va">new_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">Banksy_smooth</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="st">'2'</span>, <span class="st">'2'</span>, <span class="st">'Other'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">metadata2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/expand.grid.html" class="external-link">expand.grid</a></span><span class="op">(</span>sample_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">sample_id</span><span class="op">)</span>,</span>
<span>                         cluster <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">new_cluster</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="co"># extract time point as 'day' from sample_id </span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>        day <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"DPI.*"</span>, <span class="st">""</span>, <span class="va">sample_id</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        rep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">".*_"</span>, <span class="st">""</span>, <span class="va">sample_id</span><span class="op">)</span><span class="op">)</span> </span>
<span>      <span class="op">)</span></span></code></pre></div>
<p>Create a single-cluster design.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># design model for testing the cluster 2</span></span>
<span><span class="va">design_model2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">cluster</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/splines/ns.html" class="external-link">ns</a></span><span class="op">(</span><span class="va">day</span>, df <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>,</span>
<span>                              data <span class="op">=</span> <span class="va">metadata2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">design_model2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">metadata2</span><span class="op">$</span><span class="va">sample_id</span>, <span class="st">"_"</span>,</span>
<span>                                  <span class="va">metadata2</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span>
<span><span class="va">design_model2</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           (Intercept) clusterOther ns(day, df = 2)1 ns(day, df = 2)2</span></span>
<span><span class="co">## 2DPI_1_2            1            0        0.0000000        0.0000000</span></span>
<span><span class="co">## 2DPI_2_2            1            0        0.0000000        0.0000000</span></span>
<span><span class="co">## 10DPI_1_2           1            0        0.5513298       -0.2274421</span></span>
<span><span class="co">##           clusterOther:ns(day, df = 2)1 clusterOther:ns(day, df = 2)2</span></span>
<span><span class="co">## 2DPI_1_2                              0                             0</span></span>
<span><span class="co">## 2DPI_2_2                              0                             0</span></span>
<span><span class="co">## 10DPI_1_2                             0                             0</span></span></code></pre>
<p>Fit the single-cluster model via <code>dsp_test</code>.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span><span class="op">$</span><span class="va">cluster2</span> <span class="op">&lt;-</span> <span class="va">new_cluster</span></span>
<span><span class="va">results2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dsp_test.html">dsp_test</a></span><span class="op">(</span><span class="va">spe</span>,</span>
<span>                    design <span class="op">=</span> <span class="va">design_model2</span>,</span>
<span>                    cluster_col <span class="op">=</span> <span class="st">"cluster2"</span>,</span>
<span>                    sample_col <span class="op">=</span> <span class="va">sample_col</span>,</span>
<span>                    condition_col <span class="op">=</span> <span class="va">condition_col</span>,</span>
<span>                    verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># count significant DSP genes (at 5% FDR significance level)</span></span>
<span><span class="va">res_global2</span> <span class="op">&lt;-</span> <span class="va">results2</span><span class="op">$</span><span class="va">gene_results</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">res_global2</span><span class="op">$</span><span class="va">FDR</span> <span class="op">&lt;=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## FALSE  TRUE </span></span>
<span><span class="co">##  4955    45</span></span></code></pre>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># identify the top DSP for cluster 2</span></span>
<span><span class="op">(</span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">results2</span><span class="op">$</span><span class="va">gene_results</span><span class="op">$</span><span class="va">gene_id</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "AMEX60DD002984"</span></span></code></pre>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract the gene_name by matching the gene_id</span></span>
<span><span class="op">(</span><span class="va">feature_name</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">gene_id</span><span class="op">[</span></span>
<span>  <span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">gene_name</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">feature</span></span>
<span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "NEFM"</span></span></code></pre>
<div class="section level4">
<h4 class="tabset" id="visualization-1">Visualization<a class="anchor" aria-label="anchor" href="#visualization-1"></a>
</h4>
<div class="section level5">
<h5 id="predicted-trend">Predicted trend<a class="anchor" aria-label="anchor" href="#predicted-trend"></a>
</h5>
<p>To explore predicted counts based on estimated coefficients, we
calculate and visualize the fitted values for NEFM. The expression of
NEFM in cluster 2 first increase and then decrease, while in the
remaining regions, the expression slightly increase over time.</p>
<details><summary>
Code
</summary><div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fitted_values</span> <span class="op">&lt;-</span> <span class="va">results2</span><span class="op">[[</span><span class="st">"glmFit"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"fitted.values"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html" class="external-link">melt</a></span><span class="op">(</span><span class="va">fitted_values</span><span class="op">[</span><span class="va">feature</span>,<span class="op">]</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span><span class="st">"row_name_column"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample_id"</span>, <span class="st">"fitted"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>        day <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"DPI.*"</span>, <span class="st">""</span>, <span class="va">sample_id</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        cluster <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">".*_"</span>, <span class="st">""</span>, <span class="va">sample_id</span><span class="op">)</span><span class="op">)</span> </span>
<span>      <span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   sample_id    fitted day cluster</span></span>
<span><span class="co">## 1  2DPI_1_2  151.7510   2       2</span></span>
<span><span class="co">## 2  2DPI_2_2  130.8886   2       2</span></span>
<span><span class="co">## 3 10DPI_1_2 1024.0528  10       2</span></span></code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">m</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">day</span>, y<span class="op">=</span><span class="va">fitted</span>, group<span class="op">=</span><span class="va">cluster</span>, colour <span class="op">=</span> <span class="va">cluster</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span>, width <span class="op">=</span> <span class="fl">0.2</span>, height <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_sqrt</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="va">feature_name</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">10</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Days post injury"</span><span class="op">)</span></span></code></pre></div>
</details><div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># figure</span></span>
<span><span class="va">plt</span></span></code></pre></div>
<p><img src="DSP_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
</div>
<div class="section level5">
<h5 id="spatial-expression-1">Spatial expression<a class="anchor" aria-label="anchor" href="#spatial-expression-1"></a>
</h5>
<p>Visualize the expression of the top gene, NEFM, across samples. By
using <code>annotation_cluster = TRUE</code>, cluster annotations are
displayed on the expression plots.</p>
<details><summary>
Code
</summary><div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">sample_ids</span>, <span class="kw">function</span><span class="op">(</span><span class="va">sample_id</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># Subset spe for each sample</span></span>
<span>    <span class="va">spe_j</span> <span class="op">&lt;-</span> <span class="va">spe</span><span class="op">[</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">==</span> <span class="va">sample_id</span><span class="op">]</span></span>
<span>    <span class="co"># Create FeaturePlot for the sample</span></span>
<span>    <span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">spe_j</span>, feature <span class="op">=</span> <span class="va">feature</span>, </span>
<span>                        cluster_col <span class="op">=</span> <span class="va">spatial_cluster</span>,</span>
<span>                        coordinates <span class="op">=</span> <span class="va">coordinates</span>, </span>
<span>                        platform <span class="op">=</span> <span class="st">"Stereo-seq"</span>,</span>
<span>                        point_size <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>                        diverging <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                        annotation_cluster <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                        annotation_title <span class="op">=</span> <span class="va">sample_id</span><span class="op">)</span></span>
<span>  </span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">plot</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">combined_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plots</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="co"># common legend</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">'collect'</span><span class="op">)</span>  </span></code></pre></div>
</details><p>The trend aligns with the model’s prediction: gene abundance in
cluster 2 peaks at 10 DPI compared to other clusters.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">combined_plot</span></span></code></pre></div>
<p><img src="DSP_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.5.2 RC (2025-10-27 r88973)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] splines   stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] edgeR_4.8.0                 limma_3.66.0               </span></span>
<span><span class="co">##  [3] patchwork_1.3.2             lubridate_1.9.4            </span></span>
<span><span class="co">##  [5] forcats_1.0.1               stringr_1.5.2              </span></span>
<span><span class="co">##  [7] dplyr_1.1.4                 purrr_1.1.0                </span></span>
<span><span class="co">##  [9] readr_2.1.5                 tidyr_1.3.1                </span></span>
<span><span class="co">## [11] tibble_3.3.0                tidyverse_2.0.0            </span></span>
<span><span class="co">## [13] reshape2_1.4.4              muSpaData_1.1.0            </span></span>
<span><span class="co">## [15] ExperimentHub_3.0.0         AnnotationHub_4.0.0        </span></span>
<span><span class="co">## [17] BiocFileCache_3.0.0         dbplyr_2.5.1               </span></span>
<span><span class="co">## [19] SpatialExperiment_1.20.0    SingleCellExperiment_1.32.0</span></span>
<span><span class="co">## [21] SummarizedExperiment_1.40.0 Biobase_2.70.0             </span></span>
<span><span class="co">## [23] GenomicRanges_1.62.0        Seqinfo_1.0.0              </span></span>
<span><span class="co">## [25] IRanges_2.44.0              S4Vectors_0.48.0           </span></span>
<span><span class="co">## [27] BiocGenerics_0.56.0         generics_0.1.4             </span></span>
<span><span class="co">## [29] MatrixGenerics_1.22.0       matrixStats_1.5.0          </span></span>
<span><span class="co">## [31] ggplot2_4.0.0               DESpace_2.0.0              </span></span>
<span><span class="co">## [33] BiocStyle_2.38.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3     jsonlite_2.0.0         magrittr_2.0.4        </span></span>
<span><span class="co">##   [4] spatstat.utils_3.2-0   magick_2.9.0           farver_2.1.2          </span></span>
<span><span class="co">##   [7] rmarkdown_2.30         fs_1.6.6               ragg_1.5.0            </span></span>
<span><span class="co">##  [10] vctrs_0.6.5            memoise_2.0.1          spatstat.explore_3.5-3</span></span>
<span><span class="co">##  [13] terra_1.8-70           htmltools_0.5.8.1      S4Arrays_1.10.0       </span></span>
<span><span class="co">##  [16] curl_7.0.0             SparseArray_1.10.0     sass_0.4.10           </span></span>
<span><span class="co">##  [19] KernSmooth_2.23-26     bslib_0.9.0            htmlwidgets_1.6.4     </span></span>
<span><span class="co">##  [22] desc_1.4.3             plyr_1.8.9             httr2_1.2.1           </span></span>
<span><span class="co">##  [25] cachem_1.1.0           lifecycle_1.0.4        pkgconfig_2.0.3       </span></span>
<span><span class="co">##  [28] Matrix_1.7-4           R6_2.6.1               fastmap_1.2.0         </span></span>
<span><span class="co">##  [31] digest_0.6.37          ggnewscale_0.5.2       AnnotationDbi_1.72.0  </span></span>
<span><span class="co">##  [34] tensor_1.5.1           textshaping_1.0.4      RSQLite_2.4.3         </span></span>
<span><span class="co">##  [37] beachmat_2.26.0        labeling_0.4.3         filelock_1.0.3        </span></span>
<span><span class="co">##  [40] timechange_0.3.0       spatstat.sparse_3.1-0  httr_1.4.7            </span></span>
<span><span class="co">##  [43] polyclip_1.10-7        abind_1.4-8            compiler_4.5.2        </span></span>
<span><span class="co">##  [46] proxy_0.4-27           bit64_4.6.0-1          withr_3.0.2           </span></span>
<span><span class="co">##  [49] S7_0.2.0               BiocParallel_1.44.0    DBI_1.2.3             </span></span>
<span><span class="co">##  [52] ggforce_0.5.0          MASS_7.3-65            rappdirs_0.3.3        </span></span>
<span><span class="co">##  [55] DelayedArray_0.36.0    rjson_0.2.23           classInt_0.4-11       </span></span>
<span><span class="co">##  [58] tools_4.5.2            units_1.0-0            goftest_1.2-3         </span></span>
<span><span class="co">##  [61] glue_1.8.0             nlme_3.1-168           grid_4.5.2            </span></span>
<span><span class="co">##  [64] sf_1.0-21              gtable_0.3.6           spatstat.data_3.1-9   </span></span>
<span><span class="co">##  [67] tzdb_0.5.0             class_7.3-23           hms_1.1.4             </span></span>
<span><span class="co">##  [70] data.table_1.17.8      XVector_0.50.0         spatstat.geom_3.6-0   </span></span>
<span><span class="co">##  [73] BiocVersion_3.22.0     pillar_1.11.1          tweenr_2.0.3          </span></span>
<span><span class="co">##  [76] lattice_0.22-7         bit_4.6.0              deldir_2.0-4          </span></span>
<span><span class="co">##  [79] tidyselect_1.2.1       locfit_1.5-9.12        Biostrings_2.78.0     </span></span>
<span><span class="co">##  [82] scuttle_1.19.0         knitr_1.50             bookdown_0.45         </span></span>
<span><span class="co">##  [85] xfun_0.53              statmod_1.5.1          stringi_1.8.7         </span></span>
<span><span class="co">##  [88] yaml_2.3.10            evaluate_1.0.5         codetools_0.2-20      </span></span>
<span><span class="co">##  [91] BiocManager_1.30.26    cli_3.6.5              systemfonts_1.3.1     </span></span>
<span><span class="co">##  [94] jquerylib_0.1.4        Rcpp_1.1.0             spatstat.random_3.4-2 </span></span>
<span><span class="co">##  [97] png_0.1-8              spatstat.univar_3.1-4  parallel_4.5.2        </span></span>
<span><span class="co">## [100] pkgdown_2.1.3          assertthat_0.2.1       blob_1.2.4            </span></span>
<span><span class="co">## [103] scales_1.4.0           e1071_1.7-16           crayon_1.5.3          </span></span>
<span><span class="co">## [106] rlang_1.1.6            KEGGREST_1.50.0</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-edgeR" class="csl-entry">
Robinson, Mark D, Davis J McCarthy, and Gordon K Smyth. 2010.
<span>“edgeR: A Bioconductor Package for Differential Expression
Analysis of Digital Gene Expression Data.”</span>
<em>Bioinformatics</em> 26 (1): 139–40. <a href="https://doi.org/10.1093/bioinformatics/btp616" class="external-link">https://doi.org/10.1093/bioinformatics/btp616</a>.
</div>
<div id="ref-Banksy" class="csl-entry">
Singhal, Vipul, Nigel Chou, Joseph Lee, Yifei Yue, Jinyue Liu, Wan Kee
Chock, Li Lin, et al. 2024. <span>“BANKSY Unifies Cell Typing and Tissue
Domain Segmentation for Scalable Spatial Omics Data Analysis.”</span>
<em>Nature Genetics</em> 56 (3): 431–41. <a href="https://doi.org/10.1038/s41588-024-01664-3" class="external-link">https://doi.org/10.1038/s41588-024-01664-3</a>.
</div>
<div id="ref-ARTISTA" class="csl-entry">
Wei, Xiaoyu, Sulei Fu, Hanbo Li, Yang Liu, Shuai Wang, Weimin Feng,
Yunzhi Yang, et al. 2022. <span>“Single-Cell Stereo-Seq Reveals Induced
Progenitor Cells Involved in Axolotl Brain Regeneration.”</span>
<em>Science</em> 377 (6610): eabp9444. <a href="https://doi.org/10.1126/science.abp9444" class="external-link">https://doi.org/10.1126/science.abp9444</a>.
</div>
<div id="ref-BayesSpace" class="csl-entry">
Zhao, Edward, Matthew R Stone, Xing Ren, Jamie Guenthoer, Kimberly S
Smythe, Thomas Pulliam, Stephen R Williams, et al. 2021. <span>“Spatial
Transcriptomics at Subspot Resolution with
<span>B</span>ayes<span>S</span>pace.”</span> <em>Nature
Biotechnology</em> 39 (11): 1375–84. <a href="https://doi.org/10.1038/s41587-021-00935-2" class="external-link">https://doi.org/10.1038/s41587-021-00935-2</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Peiying Cai, Simone Tiberi.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
