<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>A framework to discover Spatially Variable genes via spatial clusters • DESpace</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="A framework to discover Spatially Variable genes via spatial clusters">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">DESpace</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.99.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/DSP.html">Differential Spatial Pattern between conditions</a></li>
    <li><a class="dropdown-item" href="../articles/SVG.html">A framework to discover Spatially Variable genes via spatial clusters</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/peicai/DESpace/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>A framework to discover Spatially Variable genes via spatial clusters</h1>
                        <h4 data-toc-skip class="author">Peiying
Cai</h4>
            <address class="author_afil">
      Institute for Molecular Life Sciences, University of Zurich,
SwitzerlandSIB Swiss Institute of Bioinformatics, University of Zurich,
Switzerland<br><a class="author_email" href="mailto:#"></a><a href="mailto:peiying.cai@uzh.ch" class="email">peiying.cai@uzh.ch</a>
      </address>
                              <h4 data-toc-skip class="author">Simone
Tiberi</h4>
            <address class="author_afil">
      Departiment of Statistical Sciences, University of Bologna,
Italy<br><a class="author_email" href="mailto:#"></a><a href="mailto:simone.tiberi@unibo.it" class="email">simone.tiberi@unibo.it</a>
      </address>
                  
            <h4 data-toc-skip class="date">03/25/2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/peicai/DESpace/blob/main/vignettes/SVG.Rmd" class="external-link"><code>vignettes/SVG.Rmd</code></a></small>
      <div class="d-none name"><code>SVG.Rmd</code></div>
    </div>

    
    
<hr>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><em>DESpace</em> is an intuitive framework for identifying spatially
variable (SV) genes (SVGs) via <em>edgeR</em> <span class="citation">(Robinson, McCarthy, and Smyth 2010)</span>, one of the
most common methods for performing differential expression analyses.</p>
<p>Based on pre-annotated spatial clusters as summarized spatial
information, <em>DESpace</em> models gene expression using a negative
binomial (NB), via <em>edgeR</em> <span class="citation">(Robinson,
McCarthy, and Smyth 2010)</span>, with spatial clusters as covariates.
SV genes (SVGs) are then identified by testing the significance of
spatial clusters.</p>
<p>Our approach assumes that the spatial structure can be summarized by
spatial clusters, which should reproduce the key features of the tissue
(e.g., white matter and layers in brain cortex). A significant test of
these covariates indicates that space influences gene expression, hence
identifying spatially variable genes.</p>
<p>Our model is flexible and robust, and is significantly faster than
the most SV methods. Furthermore, to the best of our knowledge, it is
the only SV approach that allows: - performing a SV test on each
individual spatial cluster, hence identifying the key regions affected
by spatial variability; - jointly fitting multiple samples, targeting
genes with consistent spatial patterns across biological replicates.</p>
<p>Below, we illustrate en example usage of the package.</p>
</div>
<div class="section level2">
<h2 id="basics">Basics<a class="anchor" aria-label="anchor" href="#basics"></a>
</h2>
<p><code>DESpace</code> is implemented as a R package within
Bioconductor, which is the main venue for omics analyses, and we use
various other Bioconductor packages (e.g., SpatialLIBD, and edgeR).</p>
<p><code>DESpace</code> package is available on Bioconductor and can be
installed with the following command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"DESpace"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Check that you have a valid Bioconductor installation</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/valid.html" class="external-link">valid</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>The development version of <code>DESpace</code>can also be installed
from the Bioconductor-devel branch or from GitHub.</p>
<p>To access the R code used in the vignettes, type:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/browseVignettes.html" class="external-link">browseVignettes</a></span><span class="op">(</span><span class="st">"DESpace"</span><span class="op">)</span></span></code></pre></div>
<p>Questions relative to <em>DESpace</em> should be reported as a new
issue at <a href="https://github.com/peicai/DESpace/issues" class="external-link"><em>BugReports</em></a>.</p>
<p>To cite <em>DESpace</em>, type:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/citation.html" class="external-link">citation</a></span><span class="op">(</span><span class="st">"DESpace"</span><span class="op">)</span></span></code></pre></div>
<p>Load R packages:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/peicai/DESpace" class="external-link">DESpace</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggforce.data-imaginist.com" class="external-link">ggforce</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a>
</h2>
<p>As an example dataset, we consider a human dorsolateral pre-frontal
cortex (DLPFC) spatial transcriptomics dataset from the 10x Genomics
Visium platform, including three neurotypical adult donors (i.e.,
biological replicates), with four images per subject <span class="citation">(Maynard et al. 2020)</span>. The full dataset consists
of 12 samples, which can be accessed via <a href="https://bioconductor.org/packages/release/data/experiment/vignettes/spatialLIBD/inst/doc/spatialLIBD.html" class="external-link"><em>spatialLIBD</em></a>
Bioconductor package.</p>
<div class="section level3">
<h3 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h3>
<p>Here, we consider a subset of the original data, consisting of three
biological replicates: 1 image for each of the three brain subjects.
Initially, in Section 3 <em>individual sample</em> , we fit our approach
on a single sample, whose data is stored in <code>spe3</code> whereas
all 3 samples will later be jointly used in Section 4 <em>Multiple
samples</em>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Connect to ExperimentHub</span></span>
<span><span class="va">ehub</span> <span class="op">&lt;-</span> <span class="fu">ExperimentHub</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html" class="external-link">ExperimentHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># Download the full real data (about 2.1 GB in RAM) use:</span></span>
<span><span class="va">spe_all</span> <span class="op">&lt;-</span> <span class="fu">spatialLIBD</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/spatialLIBD/man/fetch_data.html" class="external-link">fetch_data</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"spe"</span>, eh <span class="op">=</span> <span class="va">ehub</span><span class="op">)</span></span>
<span><span class="co"># Specify column names of spatial coordinates in colData(spe) </span></span>
<span><span class="va">coordinates</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"array_row"</span>, <span class="st">"array_col"</span><span class="op">)</span></span>
<span><span class="co"># Specify column names of spatial clusters in colData(spe) </span></span>
<span><span class="va">cluster_col</span> <span class="op">&lt;-</span> <span class="st">'layer_guess_reordered'</span></span>
<span><span class="co"># Remove spots missing annotations</span></span>
<span><span class="va">spe_all</span> <span class="op">&lt;-</span> <span class="va">spe_all</span><span class="op">[</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">spe_all</span><span class="op">[[</span><span class="va">cluster_col</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co"># Create three spe objects, one per sample:</span></span>
<span><span class="va">spe1</span> <span class="op">&lt;-</span> <span class="va">spe_all</span><span class="op">[</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">spe_all</span><span class="op">)</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">'151507'</span><span class="op">]</span></span>
<span><span class="va">spe2</span> <span class="op">&lt;-</span> <span class="va">spe_all</span><span class="op">[</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">spe_all</span><span class="op">)</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">'151669'</span><span class="op">]</span></span>
<span><span class="va">spe3</span> <span class="op">&lt;-</span> <span class="va">spe_all</span><span class="op">[</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">spe_all</span><span class="op">)</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">'151673'</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">spe_all</span><span class="op">)</span></span>
<span><span class="co"># Select small set of random genes for faster runtime in this example</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">sel_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">spe1</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,<span class="fl">2000</span><span class="op">)</span></span>
<span><span class="va">spe1</span> <span class="op">&lt;-</span> <span class="va">spe1</span><span class="op">[</span><span class="va">sel_genes</span>,<span class="op">]</span></span>
<span><span class="va">spe2</span> <span class="op">&lt;-</span> <span class="va">spe2</span><span class="op">[</span><span class="va">sel_genes</span>,<span class="op">]</span></span>
<span><span class="va">spe3</span> <span class="op">&lt;-</span> <span class="va">spe3</span><span class="op">[</span><span class="va">sel_genes</span>,<span class="op">]</span></span>
<span><span class="co"># For covenience, we use “gene names” instead of “gene ids”:</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">spe1</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">spe1</span><span class="op">)</span><span class="op">$</span><span class="va">gene_name</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">spe2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">spe2</span><span class="op">)</span><span class="op">$</span><span class="va">gene_name</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">spe3</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">spe3</span><span class="op">)</span><span class="op">$</span><span class="va">gene_name</span></span></code></pre></div>
<p>The spatial tissues of each sample were manually annotated in the
original manuscript <span class="citation">(Maynard et al. 2020)</span>,
and spots were labeled into one of the following categories: white
matter (WM) and layers 1 to 6. The manual annotations are stored in
column <code>layer_guess_reordered</code> of the <code>colData</code>,
while columns <code>array_col</code> and <code>array_row</code> provide
the spatial coordinates of spots.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We select a subset of columns</span></span>
<span><span class="va">keep_col</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">coordinates</span>,<span class="va">cluster_col</span>,<span class="st">"expr_chrM_ratio"</span>,<span class="st">"cell_count"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe3</span><span class="op">)</span><span class="op">[</span><span class="va">keep_col</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 5 columns</span></span>
<span><span class="co">##                    array_row array_col layer_guess_reordered expr_chrM_ratio</span></span>
<span><span class="co">##                    &lt;integer&gt; &lt;integer&gt;              &lt;factor&gt;       &lt;numeric&gt;</span></span>
<span><span class="co">## AAACAAGTATCTCCCA-1        50       102                Layer3        0.166351</span></span>
<span><span class="co">## AAACAATCTACTAGCA-1         3        43                Layer1        0.122376</span></span>
<span><span class="co">## AAACACCAATAACTGC-1        59        19                WM            0.114089</span></span>
<span><span class="co">## AAACAGAGCGACTCCT-1        14        94                Layer3        0.242223</span></span>
<span><span class="co">## AAACAGCTTTCAGAAG-1        43         9                Layer5        0.152174</span></span>
<span><span class="co">## AAACAGGGTCTATATT-1        47        13                Layer6        0.155095</span></span>
<span><span class="co">##                    cell_count</span></span>
<span><span class="co">##                     &lt;integer&gt;</span></span>
<span><span class="co">## AAACAAGTATCTCCCA-1          6</span></span>
<span><span class="co">## AAACAATCTACTAGCA-1         16</span></span>
<span><span class="co">## AAACACCAATAACTGC-1          5</span></span>
<span><span class="co">## AAACAGAGCGACTCCT-1          2</span></span>
<span><span class="co">## AAACAGCTTTCAGAAG-1          4</span></span>
<span><span class="co">## AAACAGGGTCTATATT-1          6</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="quality-controlfiltering">Quality control/filtering<a class="anchor" aria-label="anchor" href="#quality-controlfiltering"></a>
</h3>
<p>Quality control (QC) procedures at the spot and gene level aim to
remove both low-quality spots, and lowly abundant genes. For QC, we
adhere to the instructions from “Orchestrating Spatially Resolved
Transcriptomics Analysis with Bioconductor” (<a href="https://lmweber.org/OSTA-book/quality-control.html" class="external-link"><em>OSTA</em></a>).
The library size, UMI counts, ratio of mitochondrial chromosome (chM)
expression, and number of cells per spot are used to identify
low-quality spots.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Sample 1:</span></span>
<span><span class="co"># Calculate per-spot QC metrics and store in colData</span></span>
<span><span class="va">spe1</span> <span class="op">&lt;-</span> <span class="fu">scuttle</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQC</a></span><span class="op">(</span><span class="va">spe1</span>,<span class="op">)</span></span>
<span><span class="co"># Remove combined set of low-quality spots</span></span>
<span><span class="va">spe1</span> <span class="op">&lt;-</span> <span class="va">spe1</span><span class="op">[</span>, <span class="op">!</span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe1</span><span class="op">)</span><span class="op">$</span><span class="va">sum</span> <span class="op">&lt;</span> <span class="fl">10</span> <span class="op">|</span>             <span class="co"># library size</span></span>
<span>                <span class="fu">colData</span><span class="op">(</span><span class="va">spe1</span><span class="op">)</span><span class="op">$</span><span class="va">detected</span> <span class="op">&lt;</span> <span class="fl">10</span> <span class="op">|</span>         <span class="co"># number of expressed genes</span></span>
<span>                <span class="fu">colData</span><span class="op">(</span><span class="va">spe1</span><span class="op">)</span><span class="op">$</span><span class="va">expr_chrM_ratio</span> <span class="op">&gt;</span> <span class="fl">0.30</span><span class="op">|</span> <span class="co"># mitochondrial expression ratio</span></span>
<span>                <span class="fu">colData</span><span class="op">(</span><span class="va">spe1</span><span class="op">)</span><span class="op">$</span><span class="va">cell_count</span> <span class="op">&gt;</span> <span class="fl">10</span><span class="op">)</span><span class="op">]</span>       <span class="co"># number of cells per spot</span></span>
<span><span class="co"># Sample 2:</span></span>
<span><span class="co"># Calculate per-spot QC metrics and store in colData</span></span>
<span><span class="va">spe2</span> <span class="op">&lt;-</span> <span class="fu">scuttle</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQC</a></span><span class="op">(</span><span class="va">spe2</span>,<span class="op">)</span></span>
<span><span class="co"># Remove combined set of low-quality spots</span></span>
<span><span class="va">spe2</span> <span class="op">&lt;-</span> <span class="va">spe2</span><span class="op">[</span>, <span class="op">!</span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe2</span><span class="op">)</span><span class="op">$</span><span class="va">sum</span> <span class="op">&lt;</span> <span class="fl">20</span> <span class="op">|</span></span>
<span>                <span class="fu">colData</span><span class="op">(</span><span class="va">spe2</span><span class="op">)</span><span class="op">$</span><span class="va">detected</span> <span class="op">&lt;</span> <span class="fl">15</span> <span class="op">|</span></span>
<span>                <span class="fu">colData</span><span class="op">(</span><span class="va">spe2</span><span class="op">)</span><span class="op">$</span><span class="va">expr_chrM_ratio</span> <span class="op">&gt;</span> <span class="fl">0.35</span><span class="op">|</span></span>
<span>                <span class="fu">colData</span><span class="op">(</span><span class="va">spe2</span><span class="op">)</span><span class="op">$</span><span class="va">cell_count</span> <span class="op">&gt;</span> <span class="fl">8</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co"># Sample 3:</span></span>
<span><span class="va">spe3</span> <span class="op">&lt;-</span> <span class="fu">scuttle</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQC</a></span><span class="op">(</span><span class="va">spe3</span>,<span class="op">)</span></span>
<span><span class="co"># Remove combined set of low-quality spots</span></span>
<span><span class="va">spe3</span> <span class="op">&lt;-</span> <span class="va">spe3</span><span class="op">[</span>, <span class="op">!</span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe3</span><span class="op">)</span><span class="op">$</span><span class="va">sum</span> <span class="op">&lt;</span> <span class="fl">25</span> <span class="op">|</span></span>
<span>                <span class="fu">colData</span><span class="op">(</span><span class="va">spe3</span><span class="op">)</span><span class="op">$</span><span class="va">detected</span> <span class="op">&lt;</span> <span class="fl">25</span> <span class="op">|</span></span>
<span>                <span class="fu">colData</span><span class="op">(</span><span class="va">spe3</span><span class="op">)</span><span class="op">$</span><span class="va">expr_chrM_ratio</span> <span class="op">&gt;</span> <span class="fl">0.3</span><span class="op">|</span></span>
<span>                <span class="fu">colData</span><span class="op">(</span><span class="va">spe3</span><span class="op">)</span><span class="op">$</span><span class="va">cell_count</span> <span class="op">&gt;</span> <span class="fl">15</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>Then, we discard lowly abundant genes, which were detected in less
than 20 spots.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For each sample i:</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">spe_i</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/eval.html" class="external-link">eval</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/parse.html" class="external-link">parse</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"spe"</span>, <span class="va">i</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="co"># Select QC threshold for lowly expressed genes: at least 20 non-zero spots:</span></span>
<span>    <span class="va">qc_low_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu">assays</span><span class="op">(</span><span class="va">spe_i</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">20</span></span>
<span>    <span class="co"># Remove lowly abundant genes</span></span>
<span>    <span class="va">spe_i</span> <span class="op">&lt;-</span> <span class="va">spe_i</span><span class="op">[</span><span class="va">qc_low_gene</span>,<span class="op">]</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/assign.html" class="external-link">assign</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"spe"</span>, <span class="va">i</span><span class="op">)</span>, <span class="va">spe_i</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"Dimension of spe"</span>, <span class="va">i</span>, <span class="st">": "</span>, <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">spe_i</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">", "</span>, <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">spe_i</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<pre><code><span><span class="co">## Dimension of spe1: 847, 4169</span></span></code></pre>
<pre><code><span><span class="co">## Dimension of spe2: 867, 3610</span></span></code></pre>
<pre><code><span><span class="co">## Dimension of spe3: 907, 3584</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="individual-sample">Individual sample<a class="anchor" aria-label="anchor" href="#individual-sample"></a>
</h2>
<p>We fit our approach to discover SVGs in an individual sample. In
Section 4 <em>Multiple samples</em>, we will show how to jointly embed
multiple replicates.</p>
<div class="section level3">
<h3 id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h3>
<p>This framework relies on spatial clusters being accessible and
successfully summarizing the primary spatial characteristics of the
data. In most datasets, these spatial features are either accessible or
can be easily generated with spatial clustering algorithms.</p>
<div class="section level4">
<h4 id="manual-annotation">Manual annotation<a class="anchor" aria-label="anchor" href="#manual-annotation"></a>
</h4>
<p>If manual annotations are provided (e.g., annotated by a
pathologist), we can directly use those. With the <code>spe</code> or
<code>spe</code> object that contains coordinates of the spot-level
data, we can visualize spatial clusters.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View LIBD layers for one sample</span></span>
<span><span class="va">CD</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">CD</span>, </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">array_col</span>,y<span class="op">=</span><span class="va">array_row</span>, </span>
<span>    color<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">layer_guess_reordered</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_reverse</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position<span class="op">=</span><span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">""</span>, title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Manually annotated spatial clusters"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SVG_files/figure-html/view%20LIBD%20layers-1.png" width="480"></p>
</div>
<div class="section level4">
<h4 id="spatially-resolved-clustering">Spatially resolved clustering<a class="anchor" aria-label="anchor" href="#spatially-resolved-clustering"></a>
</h4>
<p>If manual annotations are not available, we can use spatially
resolved clustering tools. These methods, by jointly employing spatial
coordinates and gene expression data, enable obtaining spatial clusters.
Although, in this vignette we use pre-computed manually annotated
clusters, below we provide links to two popular spatially resolved
clustering tools: BayesSpace <span class="citation">(Zhao et al.
2021)</span> and StLearn <span class="citation">(Pham et al.
2020)</span>.</p>
<div class="section level5">
<h5 id="bayesspace">BayesSpace<a class="anchor" aria-label="anchor" href="#bayesspace"></a>
</h5>
<p>BayesSpace is a Bioconductor package that provides a Bayesian
statistical approach for spatial transcriptomics data clustering (<a href="https://edward130603.github.io/BayesSpace/index.html" class="external-link"><em>BayesSpace</em></a>).
There is a specific vignette for using BayesSpace on this dataset (human
DLPFC) <a href="https://edward130603.github.io/BayesSpace/articles/maynard_DLPFC.html" class="external-link"><em>here</em></a>.</p>
<p>After using BayesSpace, the spatial cluster assignments
(<code>spatial.cluster</code>) are available in the
<code>colData(spe)</code>.</p>
</div>
<div class="section level5">
<h5 id="stlearn">StLearn<a class="anchor" aria-label="anchor" href="#stlearn"></a>
</h5>
<p>StLearn, a python-based package, is designed for spatial
transciptomics data. It allows spatially-resolved clustering based on
Louvain or k-means (<a href="https://stlearn.readthedocs.io/en/latest/index.html" class="external-link"><em>stLearn</em></a>).
There is a tutorial for using StLearn on this dataset (human DLPFC) <a href="https://stlearn.readthedocs.io/en/latest/tutorials/stSME_clustering.html#Human-Brain-dorsolateral-prefrontal-cortex-(DLPFC)" class="external-link"><em>here</em></a>.</p>
<p>After running stLearn, we can store results as a <code>csv</code>
file.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="co"># Save spatial results</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>obsm.to_csv(<span class="st">"stLearn_clusters.csv"</span>) </span></code></pre></div>
<p>Then, we can load these results in R and store spatial clusters in
the <code>spe</code> object.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stLearn_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"stLearn_clusters.csv"</span>, sep <span class="op">=</span> <span class="st">','</span>, </span>
<span>                            header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># Match colData(spe) and stLearn results</span></span>
<span><span class="va">stLearn_results</span> <span class="op">&lt;-</span> <span class="va">stLearn_results</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe3</span><span class="op">)</span><span class="op">)</span>, </span>
<span>                                    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">stLearn_results</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">spe3</span><span class="op">)</span><span class="op">$</span><span class="va">stLearn_clusters</span> <span class="op">&lt;-</span> <span class="va">stLearn_results</span><span class="op">$</span><span class="va">stLearn_pca_kmeans</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="sv-testing">SV testing<a class="anchor" aria-label="anchor" href="#sv-testing"></a>
</h3>
<p>Once we have spatial clusters, we can search for SVGs.</p>
<div class="section level4">
<h4 id="gene-level-test">Gene-level test<a class="anchor" aria-label="anchor" href="#gene-level-test"></a>
</h4>
<p>Fit the model via <code>svg_test</code> function. Parameter
<code>spe</code> specifies the input <code>SpatialExperiment</code> or
<code>SingleCellExperiment</code> object, while <code>cluster_col</code>
defines the column names of <code>colData(spe)</code> containing spatial
clusters. To obtain all statistics, set <code>verbose</code> to
<code>TRUE</code> (default value).</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/svg_test.html">svg_test</a></span><span class="op">(</span>spe <span class="op">=</span> <span class="va">spe3</span>,</span>
<span>                        cluster_col <span class="op">=</span> <span class="va">cluster_col</span>, </span>
<span>                        verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## using 'svg_test' for spatial gene/pattern detection.</span></span></code></pre>
<pre><code><span><span class="co">## Filter low quality genes:</span></span></code></pre>
<pre><code><span><span class="co">## min_counts = 20; min_non_zero_spots = 10.</span></span></code></pre>
<pre><code><span><span class="co">## The number of genes that pass filtering is 907.</span></span></code></pre>
<pre><code><span><span class="co">## single sample test</span></span></code></pre>
<p>A list of results is returned. The main results of interest are
stored in the <code>gene_results</code>: a <code>data.fame</code>, where
columns contain gene names (<code>gene_id</code>), likelihood ratio test
statistics (<code>LR</code>), average (across spots) log-2 counts per
million (<code>logCPM</code>), raw p-values (<code>PValue</code>) and
Benjamini-Hochberg adjusted p-values (<code>FDR</code>).</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">gene_results</span>, <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         gene_id       LR   logCPM        PValue           FDR</span></span>
<span><span class="co">## SNCG       SNCG 1302.568 14.36227 3.011631e-278 2.731550e-275</span></span>
<span><span class="co">## ATP1A3   ATP1A3 1235.840 14.68931 8.376122e-264 3.798571e-261</span></span>
<span><span class="co">## PLEKHH1 PLEKHH1 1045.531 13.69919 1.268125e-222 3.833964e-220</span></span></code></pre>
<p>The second element of the results (a <code>DGEList</code> object
<code>estimated_y</code>) contains the estimated common dispersion,
which can later be used to speed-up calculation when testing individual
clusters.</p>
<p>The third and forth element of the results (<code>DGEGLM</code> and
<code>DGELRT</code> objects) contain full statistics from
<code><a href="https://rdrr.io/pkg/edgeR/man/glmfit.html" class="external-link">edgeR::glmFit</a></code> and <code><a href="https://rdrr.io/pkg/edgeR/man/glmLRT.html" class="external-link">edgeR::glmLRT</a></code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">estimated_y</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">glmLrt</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">glmFit</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "DGEList"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "edgeR"</span></span></code></pre>
<pre><code><span><span class="co">## [1] "DGELRT"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "edgeR"</span></span></code></pre>
<pre><code><span><span class="co">## [1] "DGEGLM"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "edgeR"</span></span></code></pre>
<p>Visualize the gene expression of the three most significant genes in
<code><a href="../reference/FeaturePlot.html">FeaturePlot()</a></code>. Note that the gene names in vector
<code>feature</code>, should also appear in the count matrix’s row
names. Specifying the column names of spatial coordinates of spots is
only necessary when they are not named <code>row</code> and
<code>col</code>.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">results</span><span class="op">$</span><span class="va">gene_results</span><span class="op">$</span><span class="va">gene_id</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "SNCG"    "ATP1A3"  "PLEKHH1"</span></span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">spe3</span>, <span class="va">feature</span>, </span>
<span>            coordinates <span class="op">=</span> <span class="va">coordinates</span>, </span>
<span>            ncol <span class="op">=</span> <span class="fl">3</span>, title <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="SVG_files/figure-html/top%20SVGs%20expression%20plot-1.png" width="700"></p>
<p>Additionally, function <code><a href="../reference/FeaturePlot.html">FeaturePlot()</a></code> can draw an outline
around each cluster.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">spe3</span>, <span class="va">feature</span>, </span>
<span>            coordinates <span class="op">=</span> <span class="va">coordinates</span>, </span>
<span>            annotation_cluster <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            cluster_col <span class="op">=</span> <span class="va">cluster_col</span>, </span>
<span>            cluster <span class="op">=</span> <span class="st">'all'</span>, </span>
<span>            legend_cluster <span class="op">=</span> <span class="cn">TRUE</span>, title <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="SVG_files/figure-html/expression%20plots%20all%20cluster-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="individual-cluster-test">Individual cluster test<a class="anchor" aria-label="anchor" href="#individual-cluster-test"></a>
</h4>
<p><em>DESpace</em> can also be used to reveal the specific areas of the
tissue affected by spatial variability; i.e., spatial clusters that are
particularly over/under abundant compared to the average. Function
<code><a href="../reference/individual_svg.html">individual_svg()</a></code> can be used to identify SVGs for each
individual cluster. Parameter <code>cluster_col</code> indicates the
column names of <code>colData(spe)</code> containing spatial
clusters.</p>
<p>For every spatial cluster we test, <code>edgeR</code> would normally
re-compute the dispersion estimates based on the specific design of the
test. However, this calculation represents the majority of the overall
computing time. Therefore, to speed-up calculations, we propose to use
the dispersion estimates which were previously computed for the
gene-level tests. Albeit this is an approximation, in our benchmarks, it
leads to comparable performance in terms of sensitivity and specificity.
If you want to use pre-computed gene-level dispersion estimates, set
<code>edgeR_y</code> to <code>estimated_y</code>. Alternatively, if you
want to re-compute dispersion estimates (significantly slower, but
marginally more accurate option), leave <code>edgeR_y</code> empty.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">cluster_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/individual_svg.html">individual_svg</a></span><span class="op">(</span><span class="va">spe3</span>, </span>
<span>                                    edgeR_y <span class="op">=</span> <span class="va">results</span><span class="op">$</span><span class="va">estimated_y</span>,</span>
<span>                                    cluster_col <span class="op">=</span> <span class="va">cluster_col</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Filter low quality genes:</span></span></code></pre>
<pre><code><span><span class="co">## min_counts = 20; min_non_zero_spots = 10.</span></span></code></pre>
<pre><code><span><span class="co">## The number of genes that pass filtering is907.</span></span></code></pre>
<pre><code><span><span class="co">## Pre-processing</span></span></code></pre>
<pre><code><span><span class="co">## Start modeling</span></span></code></pre>
<pre><code><span><span class="co">## Returning results</span></span></code></pre>
<p><code><a href="../reference/individual_svg.html">individual_svg()</a></code> returns a list containing the results
of the individual cluster tests. Similarly to above, for each cluster,
results are reported as a <code>data.fame</code>, where columns contain
gene names (<code>gene_id</code>), likelihood ratio test statistics
(<code>LR</code>), log2-fold changes (<code>logFC</code>), raw p-values
(<code>PValue</code>) and Benjamini-Hochberg adjusted p-values
(<code>FDR</code>). NB: that the <code>logFC</code> compares each
cluster to the rest of the tissue; e.g., a logFC of 2 for WM test
indicates that the average gene expression in WM is (4 times) higher
than the average gene expression in non-WM tissue.</p>
<p>Visualize results for WM.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">cluster_results</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "list"</span></span></code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cluster_results</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Layer1" "Layer2" "Layer3" "Layer4" "Layer5" "Layer6" "WM"</span></span></code></pre>
<p><code>top_results</code> function can be used to combine gene-and
cluster-level results. By default, results from
<code><a href="../reference/top_results.html">top_results()</a></code> report both adjusted p-values and log2-FC;
however, users can also choose to only report either, by specifying
<code>select = "FDR"</code> or <code>select = "logFC"</code>. Below,
<code>gene_PValue</code> and <code>gene_FDR</code> columns refer to the
gene-level testing, while the subsequent columns indicate the
cluster-specific results.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">merge_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/top_results.html">top_results</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">gene_results</span>, <span class="va">cluster_results</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">merge_res</span>,<span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   gene_id  gene_LR gene_logCPM   gene_Pvalue      gene_FDR   Layer1_FDR</span></span>
<span><span class="co">## 1    SNCG 1302.568    14.36227 3.011631e-278 2.731550e-275 4.639333e-13</span></span>
<span><span class="co">## 2  ATP1A3 1235.840    14.68931 8.376122e-264 3.798571e-261 3.932518e-15</span></span>
<span><span class="co">## 3 PLEKHH1 1045.531    13.69919 1.268125e-222 3.833964e-220 6.918017e-15</span></span>
<span><span class="co">##     Layer2_FDR   Layer3_FDR   Layer4_FDR   Layer5_FDR   Layer6_FDR</span></span>
<span><span class="co">## 1 2.806987e-06 8.405926e-80 7.600571e-13 2.918858e-26 1.920668e-65</span></span>
<span><span class="co">## 2 2.024988e-03 5.101748e-60 6.778389e-12 1.448295e-15 7.998865e-25</span></span>
<span><span class="co">## 3 1.210448e-06 6.445085e-44 1.058168e-06 4.480970e-17 3.190486e-01</span></span>
<span><span class="co">##          WM_FDR Layer1_logFC Layer2_logFC Layer3_logFC Layer4_logFC</span></span>
<span><span class="co">## 1 3.542972e-133   -0.7714273   -0.4604297    0.8320111    0.6135088</span></span>
<span><span class="co">## 2 2.069121e-182   -0.6342197    0.2225217    0.5509819    0.4556480</span></span>
<span><span class="co">## 3 7.833644e-209   -1.4032821   -0.7711669   -1.1047480   -0.8702769</span></span>
<span><span class="co">##   Layer5_logFC Layer6_logFC  WM_logFC</span></span>
<span><span class="co">## 1    0.5492228   -1.0925815 -1.966913</span></span>
<span><span class="co">## 2    0.3252719   -0.4761104 -1.728962</span></span>
<span><span class="co">## 3   -0.8053807   -0.1317909  2.320301</span></span></code></pre>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">merge_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/top_results.html">top_results</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">gene_results</span>, <span class="va">cluster_results</span>, </span>
<span>                        select <span class="op">=</span> <span class="st">"FDR"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">merge_res</span>,<span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   gene_id  gene_LR gene_logCPM   gene_Pvalue      gene_FDR   Layer1_FDR</span></span>
<span><span class="co">## 1    SNCG 1302.568    14.36227 3.011631e-278 2.731550e-275 4.639333e-13</span></span>
<span><span class="co">## 2  ATP1A3 1235.840    14.68931 8.376122e-264 3.798571e-261 3.932518e-15</span></span>
<span><span class="co">## 3 PLEKHH1 1045.531    13.69919 1.268125e-222 3.833964e-220 6.918017e-15</span></span>
<span><span class="co">##     Layer2_FDR   Layer3_FDR   Layer4_FDR   Layer5_FDR   Layer6_FDR</span></span>
<span><span class="co">## 1 2.806987e-06 8.405926e-80 7.600571e-13 2.918858e-26 1.920668e-65</span></span>
<span><span class="co">## 2 2.024988e-03 5.101748e-60 6.778389e-12 1.448295e-15 7.998865e-25</span></span>
<span><span class="co">## 3 1.210448e-06 6.445085e-44 1.058168e-06 4.480970e-17 3.190486e-01</span></span>
<span><span class="co">##          WM_FDR</span></span>
<span><span class="co">## 1 3.542972e-133</span></span>
<span><span class="co">## 2 2.069121e-182</span></span>
<span><span class="co">## 3 7.833644e-209</span></span></code></pre>
<p>We can further specify a cluster and check top genes detected by
<em>DESpace</em>.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check top genes for WM</span></span>
<span><span class="va">results_WM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/top_results.html">top_results</a></span><span class="op">(</span>cluster_results <span class="op">=</span> <span class="va">cluster_results</span>, </span>
<span>                        cluster <span class="op">=</span> <span class="st">"WM"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">results_WM</span>, <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         Cluster gene_id Cluster_LR Cluster_logCPM Cluster_logFC Cluster_PValue</span></span>
<span><span class="co">## PLEKHH1      WM PLEKHH1   964.6586       13.69919      2.320301  8.636874e-212</span></span>
<span><span class="co">## ATP1A3       WM  ATP1A3   841.7313       14.68931     -1.728962  4.562561e-185</span></span>
<span><span class="co">## SLAIN1       WM  SLAIN1   723.0535       13.78576      1.870661  2.901872e-159</span></span>
<span><span class="co">##           Cluster_FDR</span></span>
<span><span class="co">## PLEKHH1 7.833644e-209</span></span>
<span><span class="co">## ATP1A3  2.069121e-182</span></span>
<span><span class="co">## SLAIN1  8.773325e-157</span></span></code></pre>
<p>With <code>high_low</code> parameter, we can further filter genes to
visualize those with higher (<code>high_low = "high"</code>) or lower
(<code>high_low = "low"</code>) average abundance in the specified
cluster, compared to the average abundance in the rest of the tissue. By
default, <code>high_low = “both”</code> and all results are
provided.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results_WM_both</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/top_results.html">top_results</a></span><span class="op">(</span>cluster_results <span class="op">=</span> <span class="va">cluster_results</span>, </span>
<span>                                cluster <span class="op">=</span> <span class="st">"WM"</span>, </span>
<span>                                high_low <span class="op">=</span> <span class="st">"both"</span><span class="op">)</span></span></code></pre></div>
<p>Here we present the highly abundant cluster SVGs; i.e., SVGs with
higher expression in WM compared to the rest of the area.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">results_WM_both</span><span class="op">$</span><span class="va">high_genes</span>, <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         Cluster gene_id Cluster_LR Cluster_logCPM Cluster_logFC Cluster_PValue</span></span>
<span><span class="co">## PLEKHH1      WM PLEKHH1   964.6586       13.69919      2.320301  8.636874e-212</span></span>
<span><span class="co">## SLAIN1       WM  SLAIN1   723.0535       13.78576      1.870661  2.901872e-159</span></span>
<span><span class="co">## CMTM5        WM   CMTM5   659.0744       13.65302      2.098511  2.374763e-145</span></span>
<span><span class="co">##           Cluster_FDR</span></span>
<span><span class="co">## PLEKHH1 7.833644e-209</span></span>
<span><span class="co">## SLAIN1  8.773325e-157</span></span>
<span><span class="co">## CMTM5   4.307821e-143</span></span></code></pre>
<p>We visualize the lowly abundant cluster SVGs; i.e., SVGs with lower
expression in WM compared to the rest of the area.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">results_WM_both</span><span class="op">$</span><span class="va">low_genes</span>, <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         Cluster gene_id Cluster_LR Cluster_logCPM Cluster_logFC Cluster_PValue</span></span>
<span><span class="co">## ATP1A3       WM  ATP1A3   841.7313       14.68931     -1.728962  4.562561e-185</span></span>
<span><span class="co">## PRKAR1B      WM PRKAR1B   669.2057       14.72797     -1.505988  1.487108e-147</span></span>
<span><span class="co">## NSF          WM     NSF   628.9360       14.50629     -1.745010  8.515901e-139</span></span>
<span><span class="co">##           Cluster_FDR</span></span>
<span><span class="co">## ATP1A3  2.069121e-182</span></span>
<span><span class="co">## PRKAR1B 3.372018e-145</span></span>
<span><span class="co">## NSF     1.287320e-136</span></span></code></pre>
<p>Visualize the gene expression of the top genes for layer WM. A
cluster outline can be drawn by specifying the column names of clusters
stored in <code>colData(spe)</code> and the vector of cluster names via
<code>cluster_col</code> and <code>cluster</code>.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># SVGs with higher than average abundance in WM</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">results_WM_both</span><span class="op">$</span><span class="va">high_genes</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">spe3</span>, <span class="va">feature</span>, cluster_col <span class="op">=</span> <span class="va">cluster_col</span>, </span>
<span>            coordinates <span class="op">=</span> <span class="va">coordinates</span>, cluster <span class="op">=</span> <span class="st">'WM'</span>, </span>
<span>            legend_cluster <span class="op">=</span> <span class="cn">TRUE</span>, annotation_cluster <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>            linewidth <span class="op">=</span> <span class="fl">0.6</span>, title <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="SVG_files/figure-html/expression%20plots%20high_low-1.png" width="700"></p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># SVGs with lower than average abundance in WM</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">results_WM_both</span><span class="op">$</span><span class="va">low_genes</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">spe3</span>, <span class="va">feature</span>, cluster_col <span class="op">=</span> <span class="va">cluster_col</span>, </span>
<span>            coordinates <span class="op">=</span> <span class="va">coordinates</span>, cluster <span class="op">=</span> <span class="st">'WM'</span>, </span>
<span>            legend_cluster <span class="op">=</span> <span class="cn">TRUE</span>, annotation_cluster <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>            linewidth <span class="op">=</span> <span class="fl">0.6</span>,title <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="SVG_files/figure-html/expression%20plots%20high_low-2.png" width="700"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="multiple-samples">Multiple samples<a class="anchor" aria-label="anchor" href="#multiple-samples"></a>
</h2>
<p>If biological replicates are available, our framework allows jointly
modeling them to target SVGs with coherent spatial patterns across
samples. This approach may be particularly beneficial when data are
characterized by a large degree of (biological and technical)
variability, such as in cancer. Importantly, only genes detected (above
filtering thresholds) in all samples will be analyzed.</p>
<div class="section level3">
<h3 id="clustering-1">Clustering<a class="anchor" aria-label="anchor" href="#clustering-1"></a>
</h3>
<p>Similar to gene-level testing, the multi-sample extension requires
pre-annotated spatial clusters.</p>
<div class="section level4">
<h4 id="manual-annotation-1">Manual annotation<a class="anchor" aria-label="anchor" href="#manual-annotation-1"></a>
</h4>
<p>If the manual annotation for each sample is available, we can combine
all samples and use manual annotations directly. Note that cluster
labels must be consistent across samples (i.e., WM in sample 1 should
represent the same tissue as WM in sample 2). With the
<code>spe.combined</code> object that contains coordinates of the
spot-level data, we can visualize spatial clusters.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="co"># Use common genes</span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">spe1</span><span class="op">)</span><span class="op">)</span>; </span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">spe2</span><span class="op">)</span><span class="op">)</span>; </span>
<span><span class="va">c</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">spe3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># find vector of common genes across all samples:</span></span>
<span><span class="va">CommonGene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="va">intersect</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span>,<span class="va">c</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">spe1</span> <span class="op">&lt;-</span> <span class="va">spe1</span><span class="op">[</span><span class="va">CommonGene</span>,<span class="op">]</span></span>
<span><span class="va">spe2</span> <span class="op">&lt;-</span> <span class="va">spe2</span><span class="op">[</span><span class="va">CommonGene</span>,<span class="op">]</span></span>
<span><span class="va">spe3</span> <span class="op">&lt;-</span> <span class="va">spe3</span><span class="op">[</span><span class="va">CommonGene</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># Combine three samples</span></span>
<span><span class="va">spe.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">spe1</span>, <span class="va">spe2</span>, <span class="va">spe3</span>, deparse.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe.combined</span><span class="op">)</span><span class="op">)</span>, </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">array_col</span>, y<span class="op">=</span><span class="va">array_row</span>,</span>
<span>    color<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">layer_guess_reordered</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">sample_id</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_reverse</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position<span class="op">=</span><span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">""</span>, title <span class="op">=</span> <span class="st">"Manually annotated spatial clusters"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SVG_files/figure-html/view%20LIBD%20layers%20%5Bmulti%5D-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="spatially-resolved-multi-sample-clustering">Spatially resolved (multi-sample) clustering<a class="anchor" aria-label="anchor" href="#spatially-resolved-multi-sample-clustering"></a>
</h4>
<p>Similarly to above, if manual annotations are not available, we can
use spatially resolved clustering tools. Both BayesSpace <span class="citation">(Zhao et al. 2021)</span> and StLearn <span class="citation">(Pham et al. 2020)</span> allow jointly clustering
multiple samples. In particular each tool has a specific vignettes for
multi-testing clustering: <a href="https://edward130603.github.io/BayesSpace/articles/joint_clustering.html" class="external-link"><em>BayesSpace
vignettes</em></a>, and <a href="https://stlearn.readthedocs.io/en/latest/tutorials/Integration_multiple_datasets.html" class="external-link"><em>stLearn
vignettes</em></a>.</p>
<div class="section level5">
<h5 id="single-sample-clustering">Single sample clustering<a class="anchor" aria-label="anchor" href="#single-sample-clustering"></a>
</h5>
<p>In our benchmarks, we have noticed that, with both
<em>BayesSpace</em> and <em>StLearn</em>, joint spatial clustering of
multiple samples is more prone to failure and inaccurate results than
spatial clustering of individual samples. Therefore, if multi-sample
clustering fails, we suggest trying to cluster individual samples (as in
Section 3 <em>Individual sample</em>) and manually match cluster ids
across samples, to ensure that “cluster 1” always refers to the same
spatial region in all samples.</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="sv-testing-1">SV testing<a class="anchor" aria-label="anchor" href="#sv-testing-1"></a>
</h3>
<p>Once we have spatial clusters for multiple samples, we add them to
<code>colData(spe.combined)</code> as the column
<code>layer_guess_reordered</code> and fit the model with spatial
clusters as covariates.</p>
<div class="section level4">
<h4 id="gene-level-test-1">Gene-level test<a class="anchor" aria-label="anchor" href="#gene-level-test-1"></a>
</h4>
<p>Fit the model via <code><a href="../reference/svg_test.html">svg_test()</a></code>. Parameter <code>spe</code>
specifies the input <code>SpatialExperiment</code> or
<code>SingleCellExperiment</code> object, while <code>cluster_col</code>
and <code>sample_col</code> define the column names of
<code>colData(spe)</code> containing spatial clusters and sample ids.
With <code>replicates = TRUE</code>, we fit the multi-sample model.</p>
<p>The second element of the result (a <code>DGEList</code> object
<code>estimated_y_multi</code>) contains the estimated common dispersion
for the multi-sample case.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">multi_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/svg_test.html">svg_test</a></span><span class="op">(</span>spe <span class="op">=</span> <span class="va">spe.combined</span>,</span>
<span>                                cluster_col <span class="op">=</span> <span class="va">cluster_col</span>,</span>
<span>                                sample_col <span class="op">=</span> <span class="st">'sample_id'</span>,</span>
<span>                                replicates <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## using 'svg_test' for spatial gene/pattern detection.</span></span></code></pre>
<pre><code><span><span class="co">## Filter low quality genes:</span></span></code></pre>
<pre><code><span><span class="co">## min_counts = 20; min_non_zero_spots = 10.</span></span></code></pre>
<pre><code><span><span class="co">## The number of genes that pass filtering is 827.</span></span></code></pre>
<pre><code><span><span class="co">## multi-sample test</span></span></code></pre>
<pre><code><span><span class="co">## Repeated column names found in count matrix</span></span></code></pre>
<p>A list of results are returned. The main results of interest are
stored in the <code>gene_results</code>.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">multi_results</span><span class="op">$</span><span class="va">gene_results</span>,<span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        gene_id       LR   logCPM PValue FDR</span></span>
<span><span class="co">## HPCAL1  HPCAL1 2063.975 14.41886      0   0</span></span>
<span><span class="co">## NSF        NSF 1524.998 14.70448      0   0</span></span>
<span><span class="co">## ATP1A3  ATP1A3 1578.198 14.88558      0   0</span></span></code></pre>
<p>The second element of the results (a <code>DGEList</code> object
<code>estimated_y</code>) contains the estimated common dispersion,
which can later be used to speed-up calculation when testing individual
clusters.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">multi_results</span><span class="op">$</span><span class="va">estimated_y</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "DGEList"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "edgeR"</span></span></code></pre>
<p>For each sample, we can visualize the gene expression of the most
significant SVGs. Note that column names of spatial coordinates of spots
should be <code>row</code> and <code>col</code>.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Top three spatially variable genes</span></span>
<span><span class="va">feature</span> <span class="op">&lt;-</span> <span class="va">multi_results</span><span class="op">$</span><span class="va">gene_results</span><span class="op">$</span><span class="va">gene_id</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">]</span>; <span class="va">feature</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "HPCAL1" "NSF"    "ATP1A3"</span></span></code></pre>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Sample names</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample_id</span><span class="op">)</span>; <span class="va">samples</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "151507" "151669" "151673"</span></span></code></pre>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Use purrr::map to combine multiple figures</span></span>
<span><span class="va">spot_plots</span> <span class="op">&lt;-</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">samples</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">j</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co">## Subset spe for each sample j</span></span>
<span>    <span class="va">spe_j</span> <span class="op">&lt;-</span> <span class="va">spe.combined</span><span class="op">[</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">spe.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">==</span> <span class="va">samples</span><span class="op">[</span><span class="va">j</span><span class="op">]</span> <span class="op">]</span></span>
<span>    <span class="co">## Store three gene expression plots with gene names in `feature` for spe_j</span></span>
<span>    <span class="va">spot_plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">spe_j</span>, <span class="va">feature</span>, </span>
<span>                            coordinates <span class="op">=</span> <span class="va">coordinates</span>,</span>
<span>                            cluster_col <span class="op">=</span> <span class="va">cluster_col</span>, title <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                            annotation_cluster <span class="op">=</span> <span class="cn">TRUE</span>, legend_cluster <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">spot_plots</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">spot_plots</span>, ncol<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="SVG_files/figure-html/top%20SVGs%20expression%20plot%20%5Bmulti%5D-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="individual-cluster-test-1">Individual cluster test<a class="anchor" aria-label="anchor" href="#individual-cluster-test-1"></a>
</h4>
<p>Similarly to what shown in Section 3 <em>Individual sample</em>, our
framework can discover the key SV spatial clusters also when jointly
fitting multiple samples. For a multi-sample testing, set
<code>replicates = TRUE</code> in <code><a href="../reference/individual_svg.html">individual_svg()</a></code>.</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">cluster_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/individual_svg.html">individual_svg</a></span><span class="op">(</span><span class="va">spe.combined</span>, </span>
<span>                                edgeR_y <span class="op">=</span> <span class="va">multi_results</span><span class="op">$</span><span class="va">estimated_y</span>,</span>
<span>                                replicates <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                cluster_col <span class="op">=</span> <span class="va">cluster_col</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Filter low quality genes:</span></span></code></pre>
<pre><code><span><span class="co">## min_counts = 20; min_non_zero_spots = 10.</span></span></code></pre>
<pre><code><span><span class="co">## The number of genes that pass filtering is827.</span></span></code></pre>
<pre><code><span><span class="co">## Pre-processing</span></span></code></pre>
<pre><code><span><span class="co">## Start modeling</span></span></code></pre>
<pre><code><span><span class="co">## Returning results</span></span></code></pre>
<p><code><a href="../reference/individual_svg.html">individual_svg()</a></code> returns a list containing the results
of individual clusters, specified in <code>cluster</code> parameter. In
this case, logFC refers to the log2-FC between the average abundance,
across all samples, of a spatial cluster and the average abundance of
all remaining clusters (e.g., WM vs. non-WM tissue).</p>
<p>Visualize results for WM.</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">cluster_results</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "list"</span></span></code></pre>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cluster_results</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Layer1" "Layer2" "Layer3" "Layer4" "Layer5" "Layer6" "WM"</span></span></code></pre>
<p>As above, <code>top_results</code> function can be used to combine
gene-level and cluster-level results.</p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">merge_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/top_results.html">top_results</a></span><span class="op">(</span><span class="va">multi_results</span><span class="op">$</span><span class="va">gene_results</span>, <span class="va">cluster_results</span>, </span>
<span>                        select <span class="op">=</span> <span class="st">"FDR"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">merge_res</span>,<span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   gene_id  gene_LR gene_logCPM gene_Pvalue gene_FDR    Layer1_FDR    Layer2_FDR</span></span>
<span><span class="co">## 1  ATP1A3 1578.198    14.88558           0        0  1.983900e-93  2.062529e-02</span></span>
<span><span class="co">## 2  HPCAL1 2063.975    14.41886           0        0  4.526699e-15 1.877187e-191</span></span>
<span><span class="co">## 3     NSF 1524.998    14.70448           0        0 5.192640e-127  4.090941e-01</span></span>
<span><span class="co">##      Layer3_FDR   Layer4_FDR    Layer5_FDR   Layer6_FDR        WM_FDR</span></span>
<span><span class="co">## 1  3.676730e-16 1.916149e-23  1.722706e-46 4.968036e-03 8.763911e-194</span></span>
<span><span class="co">## 2 4.827595e-126 7.531943e-51 1.021543e-117 3.513481e-15  8.838077e-39</span></span>
<span><span class="co">## 3  4.327946e-01 1.895747e-24  3.639402e-60 1.000252e-01 3.066341e-147</span></span></code></pre>
<p>We can further select a cluster of interest, and check the top genes
detected in that cluster.</p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check top genes for WM</span></span>
<span><span class="va">results_WM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/top_results.html">top_results</a></span><span class="op">(</span>cluster_results <span class="op">=</span> <span class="va">cluster_results</span>, </span>
<span>                        cluster <span class="op">=</span> <span class="st">"WM"</span><span class="op">)</span></span>
<span><span class="co"># For each gene, adjusted p-values for each cluster</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">results_WM</span>,<span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         Cluster gene_id Cluster_LR Cluster_logCPM Cluster_logFC Cluster_PValue</span></span>
<span><span class="co">## PLEKHH1      WM PLEKHH1   933.0275       14.00544      1.725114  6.489199e-205</span></span>
<span><span class="co">## ATP1A3       WM  ATP1A3   893.8617       14.88558     -1.285887  2.119446e-196</span></span>
<span><span class="co">## SNCG         WM    SNCG   770.7951       14.60584     -1.597405  1.207471e-169</span></span>
<span><span class="co">##           Cluster_FDR</span></span>
<span><span class="co">## PLEKHH1 5.366567e-202</span></span>
<span><span class="co">## ATP1A3  8.763911e-194</span></span>
<span><span class="co">## SNCG    3.328596e-167</span></span></code></pre>
<p>With <code>high_low = "both"</code>, we can further filter genes to
visualize highly and lowly abundant SVGs.</p>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results_WM_both</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/top_results.html">top_results</a></span><span class="op">(</span>cluster_results <span class="op">=</span> <span class="va">cluster_results</span>, </span>
<span>                            cluster <span class="op">=</span> <span class="st">"WM"</span>, high_low <span class="op">=</span> <span class="st">"both"</span><span class="op">)</span></span></code></pre></div>
<p>Here we present the highly abundant cluster SVGs; i.e., SVGs with
higher expression in WM compared to the rest of the tissue.</p>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">results_WM_both</span><span class="op">$</span><span class="va">high_genes</span>,<span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         Cluster gene_id Cluster_LR Cluster_logCPM Cluster_logFC Cluster_PValue</span></span>
<span><span class="co">## PLEKHH1      WM PLEKHH1   933.0275       14.00544      1.725114  6.489199e-205</span></span>
<span><span class="co">## SLAIN1       WM  SLAIN1   660.4590       14.07042      1.407399  1.187164e-145</span></span>
<span><span class="co">## CMTM5        WM   CMTM5   617.2535       13.99119      1.479961  2.958830e-136</span></span>
<span><span class="co">##           Cluster_FDR</span></span>
<span><span class="co">## PLEKHH1 5.366567e-202</span></span>
<span><span class="co">## SLAIN1  1.636308e-143</span></span>
<span><span class="co">## CMTM5   3.495646e-134</span></span></code></pre>
<p>We visualize the lowly abundant cluster SVGs; i.e., SVGs with lower
expression in WM compared to the rest of the tissue.</p>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">results_WM_both</span><span class="op">$</span><span class="va">low_genes</span>,<span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         Cluster gene_id Cluster_LR Cluster_logCPM Cluster_logFC Cluster_PValue</span></span>
<span><span class="co">## ATP1A3       WM  ATP1A3   893.8617       14.88558     -1.285887  2.119446e-196</span></span>
<span><span class="co">## SNCG         WM    SNCG   770.7951       14.60584     -1.597405  1.207471e-169</span></span>
<span><span class="co">## PRKAR1B      WM PRKAR1B   703.9793       14.89242     -1.135692  4.077463e-155</span></span>
<span><span class="co">##           Cluster_FDR</span></span>
<span><span class="co">## ATP1A3  8.763911e-194</span></span>
<span><span class="co">## SNCG    3.328596e-167</span></span>
<span><span class="co">## PRKAR1B 8.430155e-153</span></span></code></pre>
<p>Visualize the gene expression of top three genes for layer WM.</p>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># SVGs with higher abundance in WM, than in non-WM tissue</span></span>
<span><span class="va">feature_high</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">results_WM_both</span><span class="op">$</span><span class="va">high_genes</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co"># SVGs with lower abundance in WM, than in non-WM tissue</span></span>
<span><span class="va">feature_low</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">results_WM_both</span><span class="op">$</span><span class="va">low_genes</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">plot_list_high</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>; <span class="va">plot_list_low</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## Sample names</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample_id</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">samples</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="co">## Subset spe for each sample j</span></span>
<span>    <span class="va">spe_j</span> <span class="op">&lt;-</span> <span class="va">spe.combined</span><span class="op">[</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">spe.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">==</span> <span class="va">samples</span><span class="op">[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span></span>
<span>    <span class="co">## Gene expression plots with top highly abundant cluster SVGs for spe_j</span></span>
<span>    <span class="va">plot_list_high</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">spe_j</span>, <span class="va">feature_high</span>, </span>
<span>                                coordinates <span class="op">=</span> <span class="va">coordinates</span>,</span>
<span>                                cluster_col <span class="op">=</span> <span class="va">cluster_col</span>, </span>
<span>                                linewidth <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>                                cluster <span class="op">=</span> <span class="st">'WM'</span>, annotation_cluster <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                legend_cluster <span class="op">=</span> <span class="cn">TRUE</span>, title <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    <span class="co">## Gene expression plots with top lowly abundant cluster SVGs for spe_j</span></span>
<span>    <span class="va">plot_list_low</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">spe_j</span>, <span class="va">feature_low</span>, </span>
<span>                                coordinates <span class="op">=</span> <span class="va">coordinates</span>,</span>
<span>                                cluster_col <span class="op">=</span> <span class="va">cluster_col</span>, </span>
<span>                                linewidth <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>                                cluster <span class="op">=</span> <span class="st">'WM'</span>, annotation_cluster <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                legend_cluster <span class="op">=</span> <span class="cn">TRUE</span>, title <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co"># Expression plots for SVGs with higher abundance in WM, than in non-WM tissue</span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plot_list_high</span>, ncol<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="SVG_files/figure-html/unnamed-chunk-13-1.png" width="960"></p>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Expression plots for SVGs with lower abundance in WM, than in non-WM tissue</span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plot_list_low</span>, ncol<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="SVG_files/figure-html/unnamed-chunk-13-2.png" width="960"></p>
</div>
<div class="section level4">
<h4 id="sample-specific-covariates-e-g--batch-effects">Sample-specific covariates (e.g., batch effects)<a class="anchor" aria-label="anchor" href="#sample-specific-covariates-e-g--batch-effects"></a>
</h4>
<p>If sample-specific covariates, such as batch effects, are available,
we can account for them in <em>DESpace</em>. The adjustment works as in
the <em>edgeR</em> original framework: the mean of the negative binomial
model is expressed as a function of spatial clusters, and additional
nuisance covariates; differential testing is then performed on spatial
clusters only, to identify SVGs. Note that sample-specific covariates
can be used instead of samples, but a joint modelling of both samples
and (sample-specific) covariates is not possible because the two
variables are nested.</p>
<p>To show an example application, we artificially separate samples in 2
batches:</p>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe.combined</span><span class="op">$</span><span class="va">batch_id</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">spe.combined</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"151507"</span>, <span class="st">"batch_1"</span>, <span class="st">"batch_2"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">spe.combined</span><span class="op">$</span><span class="va">batch_id</span>, <span class="va">spe.combined</span><span class="op">$</span><span class="va">sample_id</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          </span></span>
<span><span class="co">##           151507 151669 151673</span></span>
<span><span class="co">##   batch_1   4169      0      0</span></span>
<span><span class="co">##   batch_2      0   3610   3584</span></span></code></pre>
<p>Analyses are performed, as explained above, in Section 5; yet, when
running <code>svg_test</code>, we set the <code>sample_col</code> to
<code>batch_id</code>:</p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">batch_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/svg_test.html">svg_test</a></span><span class="op">(</span>spe <span class="op">=</span> <span class="va">spe.combined</span>,</span>
<span>                                cluster_col <span class="op">=</span> <span class="va">cluster_col</span>,</span>
<span>                                sample_col <span class="op">=</span> <span class="st">'batch_id'</span>,</span>
<span>                                replicates <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.5.0 alpha (2025-03-24 r88048)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] SpatialExperiment_1.17.0    SingleCellExperiment_1.29.2</span></span>
<span><span class="co">##  [3] SummarizedExperiment_1.37.0 Biobase_2.67.0             </span></span>
<span><span class="co">##  [5] GenomicRanges_1.59.1        GenomeInfoDb_1.43.4        </span></span>
<span><span class="co">##  [7] IRanges_2.41.3              S4Vectors_0.45.4           </span></span>
<span><span class="co">##  [9] BiocGenerics_0.53.6         generics_0.1.3             </span></span>
<span><span class="co">## [11] MatrixGenerics_1.19.1       matrixStats_1.5.0          </span></span>
<span><span class="co">## [13] ggforce_0.4.2               ggplot2_3.5.1              </span></span>
<span><span class="co">## [15] DESpace_1.99.2              BiocStyle_2.35.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] progress_1.2.3           tiff_0.1-12              goftest_1.2-3           </span></span>
<span><span class="co">##   [4] DT_0.33                  Biostrings_2.75.4        vctrs_0.6.5             </span></span>
<span><span class="co">##   [7] spatstat.random_3.3-3    digest_0.6.37            png_0.1-8               </span></span>
<span><span class="co">##  [10] corpcor_1.6.10           shape_1.4.6.1            proxy_0.4-27            </span></span>
<span><span class="co">##  [13] ggrepel_0.9.6            deldir_2.0-4             parallelly_1.43.0       </span></span>
<span><span class="co">##  [16] magick_2.8.6             MASS_7.3-65              pkgdown_2.1.1           </span></span>
<span><span class="co">##  [19] reshape2_1.4.4           foreach_1.5.2            httpuv_1.6.15           </span></span>
<span><span class="co">##  [22] withr_3.0.2              xfun_0.51                ggpubr_0.6.0            </span></span>
<span><span class="co">##  [25] memoise_2.0.1            benchmarkme_1.0.8        ggbeeswarm_0.7.2        </span></span>
<span><span class="co">##  [28] systemfonts_1.2.1        smoothr_1.0.1            ragg_1.3.3              </span></span>
<span><span class="co">##  [31] GlobalOptions_0.1.2      gtools_3.9.5             sosta_0.99.5            </span></span>
<span><span class="co">##  [34] V8_6.0.2                 Formula_1.2-5            prettyunits_1.2.0       </span></span>
<span><span class="co">##  [37] rematch2_2.1.2           KEGGREST_1.47.0          promises_1.3.2          </span></span>
<span><span class="co">##  [40] httr_1.4.7               rstatix_0.7.2            restfulr_0.0.15         </span></span>
<span><span class="co">##  [43] globals_0.16.3           UCSC.utils_1.3.1         units_0.8-7             </span></span>
<span><span class="co">##  [46] concaveman_1.1.0         curl_6.2.2               ScaledMatrix_1.15.0     </span></span>
<span><span class="co">##  [49] polyclip_1.10-7          GenomeInfoDbData_1.2.14  ExperimentHub_2.15.0    </span></span>
<span><span class="co">##  [52] SparseArray_1.7.7        golem_0.5.1              fftwtools_0.9-11        </span></span>
<span><span class="co">##  [55] xtable_1.8-4             stringr_1.5.1            desc_1.4.3              </span></span>
<span><span class="co">##  [58] doParallel_1.0.17        evaluate_1.0.3           S4Arrays_1.7.3          </span></span>
<span><span class="co">##  [61] BiocFileCache_2.15.1     hms_1.1.3                bookdown_0.42           </span></span>
<span><span class="co">##  [64] irlba_2.3.5.1            colorspace_2.1-1         filelock_1.0.3          </span></span>
<span><span class="co">##  [67] spatstat.data_3.1-6      shinyWidgets_0.9.0       magrittr_2.0.3          </span></span>
<span><span class="co">##  [70] later_1.4.1              viridis_0.6.5            lattice_0.22-6          </span></span>
<span><span class="co">##  [73] spatstat.geom_3.3-6      future.apply_1.11.3      XML_3.99-0.18           </span></span>
<span><span class="co">##  [76] scuttle_1.17.0           cowplot_1.1.3            class_7.3-23            </span></span>
<span><span class="co">##  [79] pillar_1.10.1            nlme_3.1-167             iterators_1.0.14        </span></span>
<span><span class="co">##  [82] EBImage_4.49.0           caTools_1.18.3           compiler_4.5.0          </span></span>
<span><span class="co">##  [85] beachmat_2.23.7          stringi_1.8.4            sf_1.0-20               </span></span>
<span><span class="co">##  [88] tensor_1.5               minqa_1.2.8              GenomicAlignments_1.43.0</span></span>
<span><span class="co">##  [91] plyr_1.8.9               crayon_1.5.3             abind_1.4-8             </span></span>
<span><span class="co">##  [94] BiocIO_1.17.1            scater_1.35.4            blme_1.0-6              </span></span>
<span><span class="co">##  [97] locfit_1.5-9.12          bit_4.6.0                terra_1.8-29            </span></span>
<span><span class="co">## [100] dplyr_1.1.4              spatialLIBD_1.19.9       codetools_0.2-20        </span></span>
<span><span class="co">## [103] textshaping_1.0.0        BiocSingular_1.23.0      bslib_0.9.0             </span></span>
<span><span class="co">## [106] e1071_1.7-16             paletteer_1.6.0          GetoptLong_1.0.5        </span></span>
<span><span class="co">## [109] plotly_4.10.4            remaCor_0.0.18           mime_0.13               </span></span>
<span><span class="co">## [112] splines_4.5.0            circlize_0.4.16          Rcpp_1.0.14             </span></span>
<span><span class="co">## [115] dbplyr_2.5.0             attempt_0.3.1            knitr_1.50              </span></span>
<span><span class="co">## [118] blob_1.2.4               clue_0.3-66              BiocVersion_3.21.1      </span></span>
<span><span class="co">## [121] lme4_1.1-36              fs_1.6.5                 listenv_0.9.1           </span></span>
<span><span class="co">## [124] Rdpack_2.6.3             ggsignif_0.6.4           tibble_3.2.1            </span></span>
<span><span class="co">## [127] Matrix_1.7-3             statmod_1.5.0            fANCOVA_0.6-1           </span></span>
<span><span class="co">## [130] tweenr_2.0.3             pkgconfig_2.0.3          tools_4.5.0             </span></span>
<span><span class="co">## [133] cachem_1.1.0             RhpcBLASctl_0.23-42      rbibutils_2.3           </span></span>
<span><span class="co">## [136] RSQLite_2.3.9            viridisLite_0.4.2        DBI_1.2.3               </span></span>
<span><span class="co">## [139] numDeriv_2016.8-1.1      fastmap_1.2.0            rmarkdown_2.29          </span></span>
<span><span class="co">## [142] scales_1.3.0             grid_4.5.0               Rsamtools_2.23.1        </span></span>
<span><span class="co">## [145] broom_1.0.7              AnnotationHub_3.15.0     sass_0.4.9              </span></span>
<span><span class="co">## [148] patchwork_1.3.0          BiocManager_1.30.25      carData_3.0-5           </span></span>
<span><span class="co">## [151] farver_2.1.2             reformulas_0.4.0         aod_1.3.3               </span></span>
<span><span class="co">## [154] mgcv_1.9-1               yaml_2.3.10              rtracklayer_1.67.1      </span></span>
<span><span class="co">## [157] cli_3.6.4                purrr_1.0.4              lifecycle_1.0.4         </span></span>
<span><span class="co">## [160] glmmTMB_1.1.10           mvtnorm_1.3-3            sessioninfo_1.2.3       </span></span>
<span><span class="co">## [163] backports_1.5.0          BiocParallel_1.41.2      gtable_0.3.6            </span></span>
<span><span class="co">## [166] rjson_0.2.23             parallel_4.5.0           limma_3.63.10           </span></span>
<span><span class="co">## [169] jsonlite_1.9.1           edgeR_4.5.9              bitops_1.0-9            </span></span>
<span><span class="co">## [172] benchmarkmeData_1.0.4    bit64_4.6.0-1            assertthat_0.2.1        </span></span>
<span><span class="co">## [175] spatstat.utils_3.1-3     BiocNeighbors_2.1.3      muscat_1.21.0           </span></span>
<span><span class="co">## [178] jquerylib_0.1.4          config_0.3.2             spatstat.univar_3.1-2   </span></span>
<span><span class="co">## [181] pbkrtest_0.5.3           lazyeval_0.2.2           shiny_1.10.0            </span></span>
<span><span class="co">## [184] htmltools_0.5.8.1        sctransform_0.4.1        rappdirs_0.3.3          </span></span>
<span><span class="co">## [187] glue_1.8.0               XVector_0.47.2           RCurl_1.98-1.17         </span></span>
<span><span class="co">## [190] classInt_0.4-11          jpeg_0.1-11              gridExtra_2.3           </span></span>
<span><span class="co">## [193] EnvStats_3.0.0           boot_1.3-31              variancePartition_1.37.2</span></span>
<span><span class="co">## [196] TMB_1.9.17               R6_2.6.1                 tidyr_1.3.1             </span></span>
<span><span class="co">## [199] DESeq2_1.47.5            gplots_3.2.0             labeling_0.4.3          </span></span>
<span><span class="co">## [202] cluster_2.1.8.1          nloptr_2.2.1             DelayedArray_0.33.6     </span></span>
<span><span class="co">## [205] tidyselect_1.2.1         vipor_0.4.7              car_3.1-3               </span></span>
<span><span class="co">## [208] AnnotationDbi_1.69.0     future_1.34.0            rsvd_1.0.5              </span></span>
<span><span class="co">## [211] munsell_0.5.1            KernSmooth_2.23-26       data.table_1.17.0       </span></span>
<span><span class="co">## [214] htmlwidgets_1.6.4        ComplexHeatmap_2.23.0    RColorBrewer_1.1-3      </span></span>
<span><span class="co">## [217] rlang_1.1.5              spatstat.sparse_3.1-0    spatstat.explore_3.4-2  </span></span>
<span><span class="co">## [220] lmerTest_3.1-3           ggnewscale_0.5.1         beeswarm_0.4.0</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-LIBD" class="csl-entry">
Maynard, Kristen R, Leonardo Collado-Torres, Lukas M Weber, Cedric
Uytingco, Brianna K Barry, Stephen R Williams, Joseph L Catallini, et
al. 2020. <span>“Transcriptome-Scale Spatial Gene Expression in the
Human Dorsolateral Prefrontal Cortex.”</span> <em>BioRxiv</em>, 2020–02.
<a href="https://doi.org/10.1038/s41593-020-00787-0" class="external-link">https://doi.org/10.1038/s41593-020-00787-0</a>.
</div>
<div id="ref-stLearn" class="csl-entry">
Pham, Duy, Xiao Tan, Jun Xu, Laura F Grice, Pui Yeng Lam, Arti Raghubar,
Jana Vukovic, Marc J Ruitenberg, and Quan Nguyen. 2020.
<span>“St<span>L</span>earn: Integrating Spatial Location, Tissue
Morphology and Gene Expression to Find Cell Types, Cell-Cell
Interactions and Spatial Trajectories Within Undissociated
Tissues.”</span> <em>Biorxiv</em>, 2020–05. <a href="https://doi.org/doi:10.1093/bioinformatics/btz914" class="external-link">https://doi.org/doi:10.1093/bioinformatics/btz914</a>.
</div>
<div id="ref-edgeR" class="csl-entry">
Robinson, Mark D, Davis J McCarthy, and Gordon K Smyth. 2010.
<span>“edgeR: A Bioconductor Package for Differential Expression
Analysis of Digital Gene Expression Data.”</span>
<em>Bioinformatics</em> 26 (1): 139–40. <a href="https://doi.org/10.1093/bioinformatics/btp616" class="external-link">https://doi.org/10.1093/bioinformatics/btp616</a>.
</div>
<div id="ref-BayesSpace" class="csl-entry">
Zhao, Edward, Matthew R Stone, Xing Ren, Jamie Guenthoer, Kimberly S
Smythe, Thomas Pulliam, Stephen R Williams, et al. 2021. <span>“Spatial
Transcriptomics at Subspot Resolution with
<span>B</span>ayes<span>S</span>pace.”</span> <em>Nature
Biotechnology</em> 39 (11): 1375–84. <a href="https://doi.org/10.1038/s41587-021-00935-2" class="external-link">https://doi.org/10.1038/s41587-021-00935-2</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Peiying Cai, Simone Tiberi.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
